// ask-cli/dist/lib/view/import-status/import-status-view-events.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.IMPORT_STATUS_CANCEL_TASK_EVENT = exports.IMPORT_STATUS_FETCHING_IMPORT_ID_SUCESS_EVENT = exports.IMPORT_STATUS_FETCHING_NEW_SKILL_ID_SUCESS_EVENT = exports.IMPORT_STATUS_FETCHING_SKILL_ID_SUCESS_EVENT = exports.IMPORT_STATUS_IM_BUILD_FAILED_EVENT = exports.IMPORT_STATUS_IM_BUILD_SUCCESS_EVENT = exports.IMPORT_STATUS_AC_BUILD_FAILED_EVENT = exports.IMPORT_STATUS_AC_BUILD_SUCCESS_EVENT = exports.IMPORT_STATUS_BUILDING_AC_FULL_EVENT = exports.IMPORT_STATUS_BUILDING_AC_LIGHT_EVENT = void 0;
exports.IMPORT_STATUS_BUILDING_AC_LIGHT_EVENT = "buildingACLight";
exports.IMPORT_STATUS_BUILDING_AC_FULL_EVENT = "buildingACFull";
exports.IMPORT_STATUS_AC_BUILD_SUCCESS_EVENT = "buildACSuccess";
exports.IMPORT_STATUS_AC_BUILD_FAILED_EVENT = "buildACFailed";
exports.IMPORT_STATUS_IM_BUILD_SUCCESS_EVENT = "buildIMSuccess";
exports.IMPORT_STATUS_IM_BUILD_FAILED_EVENT = "buildIMFailed";
exports.IMPORT_STATUS_FETCHING_SKILL_ID_SUCESS_EVENT = "fetchingSkillIdSuccess";
exports.IMPORT_STATUS_FETCHING_NEW_SKILL_ID_SUCESS_EVENT = "fetchingNewSkillIdSuccess";
exports.IMPORT_STATUS_FETCHING_IMPORT_ID_SUCESS_EVENT = "fetchingImportIdSuccess";
exports.IMPORT_STATUS_CANCEL_TASK_EVENT = "cancelTask";
