// @alexa/acdl/dist/cjs/acc/CompileZipAndDeploy.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.CompileZipAndDeploy = void 0;
/* eslint-disable no-console */
const path_1 = __importDefault(require("path"));
const fs_1 = __importDefault(require("fs"));
const validate_1 = require("../validate");
const bundle_1 = require("../bundle");
const synthesize_1 = require("../synthesize");
const project_config_1 = require("../project-config");
const project_1 = require("../project");
const error_1 = require("../error");
const smapi_1 = require("../smapi");
class CompileZipAndDeploy {
    /**
     * Compiles skill or reusable component. Will also zip the compiled ASKIR skill output if the
     * command is "package" or "deploy" and deploy the skill if the command is deploy
     * @param skipValidation true to skip validation
     * @param profileName profile name from a call to ProfileNameGetter.getProfileName
     * @param skillId skill id used if command is deploy
     * @param command compilation command which is either "" "deploy". "package" or "compile"
     */
    static async compileZipAndDeploy(skipValidation, profileName, skillId, command, askCliConfig) {
        const projectConfig = await (0, project_config_1.loadProjectConfig)(process.cwd(), profileName);
        const outDirPath = path_1.default.join(projectConfig.rootDir, projectConfig.outDir);
        const skillPackageOutputDir = path_1.default.join(outDirPath, "skill-package");
        // Remove the old generated contents
        CompileZipAndDeploy.cleanDirectory(skillPackageOutputDir);
        const project = await (0, project_1.loadProject)(projectConfig);
        const isSkill = command === "package" || command === "deploy" || project.isSkill();
        const sourceDirectory = project.config.sourceRootDir || project.config.srcDir || "";
        const absSourceDirectory = path_1.default.isAbsolute(sourceDirectory) ? sourceDirectory : path_1.default.join(process.cwd(), sourceDirectory);
        console.log(`************ Compiling ${isSkill ? "Skill" : "Module"} ************`);
        console.log(`${isSkill ? "Skill package" : "Source"} directory: ${absSourceDirectory}`);
        const errors = CompileZipAndDeploy.validateAndTerminateOnErrors(skipValidation, project, false);
        // isSkill is true for skills typically with a skill-package directory usually packaged with bundleProject
        // A module is typically a reusable component typically packaged with synthesizeProject
        if (isSkill) {
            const bundleProjectOutput = await (0, bundle_1.bundleProject)(project);
            if (!skipValidation) {
                CompileZipAndDeploy.logAndTerminateOnError(bundleProjectOutput.errors);
            }
            // The "package" command will zip the contents of the skill, deploy will both zip and deploy
            if (command === "deploy" || command === "package") {
                const zippedBundleOutput = await (0, bundle_1.zipBundleOutput)(bundleProjectOutput);
                if (command === "deploy") {
                    if (!zippedBundleOutput.skillPackageZip) {
                        throw new Error("Unable to find zip to deploy.");
                    }
                    await CompileZipAndDeploy.deploy(profileName, skillId, zippedBundleOutput.skillPackageZip, askCliConfig);
                }
            }
        }
        else {
            await (0, synthesize_1.synthesizeProject)(project);
        }
        (0, error_1.logProjectWarnings)(errors);
        console.log(`Compiled ${isSkill ? "Skill" : "Module"} Artifacts successfully,  the output is saved at ${path_1.default.join(process.cwd(), project.config.outDir)}.`);
    }
    static async initSmapi(profileName, askCliConfig) {
        const lwaProfile = askCliConfig.getLwaProfile(profileName);
        return new smapi_1.SmapiClient(lwaProfile);
    }
    /**
     * Removes a directory recursively
     * @param directoryPath path to clean
     */
    static cleanDirectory(directoryPath) {
        if (fs_1.default.existsSync(directoryPath)) {
            fs_1.default.rmSync(directoryPath, { recursive: true });
        }
    }
    /**
     * Validates the project terminating on error
     * @param skipValidation true if we should skip validations
     * @param project Project loaded with loadProject
     * @param isSkill true when validating a skill, false when validating a reusable component
     */
    static validateAndTerminateOnErrors(skipValidation, project, isSkill) {
        const errors = (0, validate_1.validateProject)(project, isSkill, skipValidation);
        CompileZipAndDeploy.logAndTerminateOnError(errors);
        return errors;
    }
    /**
     *  When there are errors (as opposed to just warnings) log the errors
     *  and terminates the process
     * @param errors errors to check
     */
    static logAndTerminateOnError(errors) {
        if (!(0, error_1.hasTerminalErrors)(errors)) {
            return;
        }
        throw errors;
    }
    /**
     * Deploys the skill
     * @param profileName profile name
     * @param skillId skill id
     * @param skillPackageZip contents to be deployed
     */
    static async deploy(profileName, skillId, skillPackageZip, askCliConfig) {
        var _a;
        const smapi = await CompileZipAndDeploy.initSmapi(profileName, askCliConfig);
        const location = await smapi.createUploadUrl();
        await smapi.uploadSkillPackage(skillPackageZip, location);
        try {
            await smapi.importSkillPackage(skillId, location);
        }
        catch (err) {
            if (typeof ((_a = err.response) === null || _a === void 0 ? void 0 : _a.message) === "string") {
                // catch errors thrown by SMAPI
                throw new Error(err.response.message);
            }
            throw err;
        }
    }
}
exports.CompileZipAndDeploy = CompileZipAndDeploy;
