// @alexa/acdl/dist/esm/acc/CompileZipAndDeploy.js
/* eslint-disable no-console */
import path from "path";
import fs from "fs";
import { validateProject } from "../validate";
import { bundleProject, zipBundleOutput } from "../bundle";
import { synthesizeProject } from "../synthesize";
import { loadProjectConfig } from "../project-config";
import { loadProject } from "../project";
import { hasTerminalErrors, logProjectWarnings } from "../error";
import { SmapiClient } from "../smapi";
export class CompileZipAndDeploy {
    /**
     * Compiles skill or reusable component. Will also zip the compiled ASKIR skill output if the
     * command is "package" or "deploy" and deploy the skill if the command is deploy
     * @param skipValidation true to skip validation
     * @param profileName profile name from a call to ProfileNameGetter.getProfileName
     * @param skillId skill id used if command is deploy
     * @param command compilation command which is either "" "deploy". "package" or "compile"
     */
    static async compileZipAndDeploy(skipValidation, profileName, skillId, command, askCliConfig) {
        const projectConfig = await loadProjectConfig(process.cwd(), profileName);
        const outDirPath = path.join(projectConfig.rootDir, projectConfig.outDir);
        const skillPackageOutputDir = path.join(outDirPath, "skill-package");
        // Remove the old generated contents
        CompileZipAndDeploy.cleanDirectory(skillPackageOutputDir);
        const project = await loadProject(projectConfig);
        const isSkill = command === "package" || command === "deploy" || project.isSkill();
        const sourceDirectory = project.config.sourceRootDir || project.config.srcDir || "";
        const absSourceDirectory = path.isAbsolute(sourceDirectory) ? sourceDirectory : path.join(process.cwd(), sourceDirectory);
        console.log(`************ Compiling ${isSkill ? "Skill" : "Module"} ************`);
        console.log(`${isSkill ? "Skill package" : "Source"} directory: ${absSourceDirectory}`);
        const errors = CompileZipAndDeploy.validateAndTerminateOnErrors(skipValidation, project, false);
        // isSkill is true for skills typically with a skill-package directory usually packaged with bundleProject
        // A module is typically a reusable component typically packaged with synthesizeProject
        if (isSkill) {
            const bundleProjectOutput = await bundleProject(project);
            if (!skipValidation) {
                CompileZipAndDeploy.logAndTerminateOnError(bundleProjectOutput.errors);
            }
            // The "package" command will zip the contents of the skill, deploy will both zip and deploy
            if (command === "deploy" || command === "package") {
                const zippedBundleOutput = await zipBundleOutput(bundleProjectOutput);
                if (command === "deploy") {
                    if (!zippedBundleOutput.skillPackageZip) {
                        throw new Error("Unable to find zip to deploy.");
                    }
                    await CompileZipAndDeploy.deploy(profileName, skillId, zippedBundleOutput.skillPackageZip, askCliConfig);
                }
            }
        }
        else {
            await synthesizeProject(project);
        }
        logProjectWarnings(errors);
        console.log(`Compiled ${isSkill ? "Skill" : "Module"} Artifacts successfully,  the output is saved at ${path.join(process.cwd(), project.config.outDir)}.`);
    }
    static async initSmapi(profileName, askCliConfig) {
        const lwaProfile = askCliConfig.getLwaProfile(profileName);
        return new SmapiClient(lwaProfile);
    }
    /**
     * Removes a directory recursively
     * @param directoryPath path to clean
     */
    static cleanDirectory(directoryPath) {
        if (fs.existsSync(directoryPath)) {
            fs.rmSync(directoryPath, { recursive: true });
        }
    }
    /**
     * Validates the project terminating on error
     * @param skipValidation true if we should skip validations
     * @param project Project loaded with loadProject
     * @param isSkill true when validating a skill, false when validating a reusable component
     */
    static validateAndTerminateOnErrors(skipValidation, project, isSkill) {
        const errors = validateProject(project, isSkill, skipValidation);
        CompileZipAndDeploy.logAndTerminateOnError(errors);
        return errors;
    }
    /**
     *  When there are errors (as opposed to just warnings) log the errors
     *  and terminates the process
     * @param errors errors to check
     */
    static logAndTerminateOnError(errors) {
        if (!hasTerminalErrors(errors)) {
            return;
        }
        throw errors;
    }
    /**
     * Deploys the skill
     * @param profileName profile name
     * @param skillId skill id
     * @param skillPackageZip contents to be deployed
     */
    static async deploy(profileName, skillId, skillPackageZip, askCliConfig) {
        var _a;
        const smapi = await CompileZipAndDeploy.initSmapi(profileName, askCliConfig);
        const location = await smapi.createUploadUrl();
        await smapi.uploadSkillPackage(skillPackageZip, location);
        try {
            await smapi.importSkillPackage(skillId, location);
        }
        catch (err) {
            if (typeof ((_a = err.response) === null || _a === void 0 ? void 0 : _a.message) === "string") {
                // catch errors thrown by SMAPI
                throw new Error(err.response.message);
            }
            throw err;
        }
    }
}
