// @alexa/acdl/dist/esm/acc/NPMPackageVersionCheck.js
import * as https from "https";
// eslint-disable-next-line import/no-extraneous-dependencies
import semver from "semver";
/**
 * Rest GET request for NPM registry
 */
const ACDL_NPM_REGISTRY_HTTPS_OPTIONS = {
    hostname: "registry.npmjs.org",
    port: 443,
    path: "/@alexa/acdl/latest",
    method: "GET",
};
/**
 * Checks the version of acc against the latest npm version producing a warning if there is
 * a new patch/minor version or an error following by termination if there is a new major version
 * to prevent backend backwards imcompatible changes that might happen for major versions and inform
 * of improvemnts in minor/patch versions
 */
export class NPMPackageVersionCheck {
    /**
     * Performs the minor/patch and major patch checks mentioned in the class description
     * @param localAccVersion local @alexa/acdl version returned when we do acc --version
     */
    static async Check(localAccVersion) {
        return new Promise((resolve) => {
            const req = https.request(ACDL_NPM_REGISTRY_HTTPS_OPTIONS, (res) => {
                let data = "";
                res.on("data", (chunk) => {
                    data += chunk;
                });
                res.on("end", () => {
                    const latestNpmVersion = JSON.parse(data).version;
                    this.checkMajorVersion(latestNpmVersion, localAccVersion);
                    this.checkMinorAndPatchVersion(latestNpmVersion, localAccVersion);
                    resolve();
                });
            });
            req.on("error", (e) => {
                console.error(e);
                resolve();
            });
            req.end();
        });
    }
    /**
     * Produces an error followed by termination if there is a newer major version
     * @param latestNpmVersion latest npm registry version of @alexa/acdl
     * @param localAccVersion local @alexa/acdl version returned when we do acc --version
     */
    static checkMajorVersion(latestNpmVersion, localAccVersion) {
        if (semver.major(localAccVersion) < semver.major(latestNpmVersion)) {
            throw new Error(`Error: You must use the latest major version of @alexa/acdl (${latestNpmVersion}) to ensure compatibility with the Alexa Conversations infrastructure.\n` +
                ` update to the latest @alexa/acdl using: 'npm install @alexa/acdl@${latestNpmVersion}'`);
        }
    }
    /**
     * Produces a warning if there is a newer minor/patch version
     * @param latestNpmVersion latest npm registry version of @alexa/acdl
     * @param localAccVersion local @alexa/acdl version returned when we do acc --version
     */
    static checkMinorAndPatchVersion(latestNpmVersion, localAccVersion) {
        if (semver.minor(localAccVersion) < semver.minor(latestNpmVersion) ||
            (semver.minor(localAccVersion) === semver.minor(latestNpmVersion) && semver.patch(localAccVersion) < semver.patch(latestNpmVersion))) {
            console.warn(`Warning: There is a new version of acc available (${latestNpmVersion}). Using the latest version might help you discover modeling issues earlier`);
            console.warn(` update to the latest @alexa/acdl using: 'npm install @alexa/acdl@${latestNpmVersion}'`);
        }
    }
}
