// @alexa/acdl/dist/cjs/ast/condition.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.When = exports.Condition = exports.ElseScope = exports.ThenScope = exports.isCondition = void 0;
const block_1 = require("./block");
const guard_1 = require("./guard");
const tree_1 = require("./tree");
/**
 * Check if a value is a Condition.
 */
exports.isCondition = (0, guard_1.guard)("Condition");
class ThenScope extends block_1.Block {
}
exports.ThenScope = ThenScope;
class ElseScope extends block_1.Block {
}
exports.ElseScope = ElseScope;
/**
 * A branch in a Condition block.
 */
class Condition extends tree_1.Tree {
    constructor(
    /**
     * Condition to branch on. Expressions are only considered if this is true.
     */
    when, 
    /**
     * Block of sample expressions for when the condition is true.
     */
    then, 
    /**
     * Block of samples to consider if the condition is false.
     */
    _else, annotations, loc) {
        super("Condition", loc);
        Object.defineProperty(this, "when", {
            enumerable: true,
            configurable: true,
            writable: true,
            value: when
        });
        Object.defineProperty(this, "then", {
            enumerable: true,
            configurable: true,
            writable: true,
            value: then
        });
        Object.defineProperty(this, "annotations", {
            enumerable: true,
            configurable: true,
            writable: true,
            value: annotations
        });
        Object.defineProperty(this, "else", {
            enumerable: true,
            configurable: true,
            writable: true,
            value: void 0
        });
        this.else = _else;
        this.setAsParentOn(when);
        this.setAsParentOn(then);
        this.setAsParentOn(_else);
        this.setAsParentOn(annotations);
    }
    clone() {
        var _a, _b, _c, _d;
        return new Condition((_a = this.when) === null || _a === void 0 ? void 0 : _a.clone(), (_b = this.then) === null || _b === void 0 ? void 0 : _b.clone(), (_c = this.else) === null || _c === void 0 ? void 0 : _c.clone(), (_d = this.annotations) === null || _d === void 0 ? void 0 : _d.map((a) => a.clone()), this.loc);
    }
}
exports.Condition = Condition;
class When extends tree_1.Tree {
    constructor(expression) {
        super("When", expression === null || expression === void 0 ? void 0 : expression.loc);
        Object.defineProperty(this, "expression", {
            enumerable: true,
            configurable: true,
            writable: true,
            value: expression
        });
        this.setAsParentOn(expression);
    }
    clone() {
        var _a;
        return new When((_a = this.expression) === null || _a === void 0 ? void 0 : _a.clone());
    }
}
exports.When = When;
