// @alexa/acdl/dist/cjs/ast/module.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.Module = exports.isModule = exports.ModuleSchema = exports.isModuleExpression = void 0;
const tree_1 = require("./tree");
const guard_1 = require("./guard");
exports.isModuleExpression = (0, guard_1.guard)("NameDeclaration", "ActionDeclaration", "DialogDeclaration", "TypeDeclaration", "InteractionDeclaration", "Call");
/**
 * JSON schema version of the Module.
 */
exports.ModuleSchema = "https://amazonalexa.com/ask/2020/12/Module";
function isModule(a) {
    return a.kind === "Module";
}
exports.isModule = isModule;
/**
 * A module contains a set of expressions.
 */
class Module extends tree_1.Tree {
    constructor(
    /**
     * URI of the file that produced this Module.
     */
    _uri, 
    /**
     * The schema for the module file.
     */
    schema, 
    /**
     * Module namespace.
     */
    namespace, 
    /**
     * Imports of the module.
     */
    imports, 
    /**
     * The declared names that are visible outside this module.
     */
    exports, 
    /**
     * The expressions in this module.
     */
    expressions, loc) {
        super("Module", loc);
        Object.defineProperty(this, "_uri", {
            enumerable: true,
            configurable: true,
            writable: true,
            value: _uri
        });
        Object.defineProperty(this, "schema", {
            enumerable: true,
            configurable: true,
            writable: true,
            value: schema
        });
        Object.defineProperty(this, "namespace", {
            enumerable: true,
            configurable: true,
            writable: true,
            value: namespace
        });
        Object.defineProperty(this, "imports", {
            enumerable: true,
            configurable: true,
            writable: true,
            value: imports
        });
        Object.defineProperty(this, "exports", {
            enumerable: true,
            configurable: true,
            writable: true,
            value: exports
        });
        Object.defineProperty(this, "expressions", {
            enumerable: true,
            configurable: true,
            writable: true,
            value: expressions
        });
        this.setAsParentOn(namespace);
        this.setAsParentOn(imports);
        this.setAsParentOn(expressions);
    }
    get uri() {
        return this._uri;
    }
    clone() {
        var _a, _b;
        return new Module(this.uri, this.schema, (_a = this.namespace) === null || _a === void 0 ? void 0 : _a.clone(), (_b = this.imports) === null || _b === void 0 ? void 0 : _b.map((i) => i.clone()), this.exports, this.expressions, this.loc);
    }
}
exports.Module = Module;
