// @alexa/acdl/dist/cjs/ast/prelude.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.Locale = exports.BuiltIns = exports.BuiltinsNamespace = exports.AlexaConversationPrompts = exports.isAnnotationTarget = exports.AnnotationTarget = exports.OPERATOR_ACTIONS_SET = exports.FunctionTypeName = exports.DialogTypeName = exports.ActionTypeName = exports.AlexaConversations = exports.AlexaConversationsNamespace = exports.DisplayNamespace = exports.PromptsNamespace = exports.SlotTypesNamespace = exports.AlexaSchema = exports.AlexaSchemaNamespace = void 0;
exports.AlexaSchemaNamespace = "com.amazon.alexa.schema";
/**
 * Statically known names of root types in Alexa schema (`@alexa/schema`)
 */
var AlexaSchema;
(function (AlexaSchema) {
    AlexaSchema["Argument"] = "com.amazon.alexa.schema.Argument";
    AlexaSchema["Arguments"] = "com.amazon.alexa.schema.Arguments";
    AlexaSchema["ArgumentProperties"] = "com.amazon.alexa.schema.ArgumentProperties";
    AlexaSchema["ArgumentProperty"] = "com.amazon.alexa.schema.ArgumentProperty";
    AlexaSchema["ArgumentValues"] = "com.amazon.alexa.schema.ArgumentValues";
    AlexaSchema["Boolean"] = "com.amazon.alexa.schema.Boolean";
    AlexaSchema["Event"] = "com.amazon.alexa.schema.Event";
    AlexaSchema["Function"] = "com.amazon.alexa.schema.Function";
    AlexaSchema["List"] = "com.amazon.alexa.schema.List";
    AlexaSchema["Nothing"] = "com.amazon.alexa.schema.Nothing";
    AlexaSchema["Number"] = "com.amazon.alexa.schema.Number";
    AlexaSchema["Optional"] = "com.amazon.alexa.schema.Optional";
    AlexaSchema["Properties"] = "com.amazon.alexa.schema.Properties";
    AlexaSchema["Property"] = "com.amazon.alexa.schema.Property";
    AlexaSchema["ReturnValue"] = "com.amazon.alexa.schema.ReturnValue";
    AlexaSchema["String"] = "com.amazon.alexa.schema.String";
    AlexaSchema["Thing"] = "com.amazon.alexa.schema.Thing";
    AlexaSchema["Time"] = "com.amazon.alexa.schema.Time";
    AlexaSchema["Tuple"] = "com.amazon.alexa.schema.Tuple";
    AlexaSchema["Type"] = "com.amazon.alexa.schema.Type";
    AlexaSchema["Void"] = "com.amazon.alexa.schema.Void";
})(AlexaSchema = exports.AlexaSchema || (exports.AlexaSchema = {}));
/**
 * Namespace given to Slot Types imported from an Interaction Model JSON file.
 */
exports.SlotTypesNamespace = "slotTypes";
/**
 * Namespace given to APL/A prompts loaded from `response/prompts`.
 */
exports.PromptsNamespace = "prompts";
/**
 * Namespace given to APL display templates loaded from `response/display`.
 */
exports.DisplayNamespace = "displays";
/**
 * Namespace of the Alexa Conversations Core Library.
 */
exports.AlexaConversationsNamespace = "com.amazon.alexa.ask.conversations";
/**
 * Statically known names of alexa conversation actions (`@alexa/ac-core`)
 */
var AlexaConversations;
(function (AlexaConversations) {
    AlexaConversations["Action"] = "com.amazon.alexa.ask.conversations.Action";
    AlexaConversations["ActionContext"] = "com.amazon.alexa.ask.conversations.ActionContext";
    AlexaConversations["ActionContextValues"] = "com.amazon.alexa.ask.conversations.ActionContextValues";
    AlexaConversations["Affirm"] = "com.amazon.alexa.ask.conversations.Affirm";
    AlexaConversations["And"] = "com.amazon.alexa.ask.conversations.and";
    AlexaConversations["annotation"] = "com.amazon.alexa.ask.conversations.annotation";
    AlexaConversations["AnnotationTarget"] = "com.amazon.alexa.ask.conversations.AnnotationTarget";
    AlexaConversations["apl"] = "com.amazon.alexa.ask.conversations.apl";
    AlexaConversations["APL"] = "com.amazon.alexa.ask.conversations.APL";
    AlexaConversations["apla"] = "com.amazon.alexa.ask.conversations.apla";
    AlexaConversations["APLA"] = "com.amazon.alexa.ask.conversations.APLA";
    AlexaConversations["Args"] = "com.amazon.alexa.ask.conversations.Args";
    AlexaConversations["Argument"] = "com.amazon.alexa.ask.conversations.Argument";
    AlexaConversations["Arguments"] = "com.amazon.alexa.ask.conversations.Arguments";
    AlexaConversations["builtinCatalog"] = "com.amazon.alexa.ask.conversations.builtinCatalog";
    AlexaConversations["Bye"] = "com.amazon.alexa.ask.conversations.Bye";
    AlexaConversations["catalog"] = "com.amazon.alexa.ask.conversations.catalog";
    AlexaConversations["Catalog"] = "com.amazon.alexa.ask.conversations.Catalog";
    AlexaConversations["CatalogItem"] = "com.amazon.alexa.ask.conversations.CatalogItem";
    AlexaConversations["confirmAction"] = "com.amazon.alexa.ask.conversations.confirmAction";
    AlexaConversations["ConfirmAction"] = "com.amazon.alexa.ask.conversations.ConfirmAction";
    AlexaConversations["confirmArgs"] = "com.amazon.alexa.ask.conversations.confirmArgs";
    AlexaConversations["ConfirmArgs"] = "com.amazon.alexa.ask.conversations.ConfirmArgs";
    AlexaConversations["ConfirmArguments"] = "com.amazon.alexa.ask.conversations.ConfirmArguments";
    AlexaConversations["delegateRequest"] = "com.amazon.alexa.ask.conversations.delegateRequest";
    AlexaConversations["delegateToIntent"] = "com.amazon.alexa.ask.conversations.delegateToIntent";
    AlexaConversations["Deny"] = "com.amazon.alexa.ask.conversations.Deny";
    AlexaConversations["Dialog"] = "com.amazon.alexa.ask.conversations.Dialog";
    AlexaConversations["ensure"] = "com.amazon.alexa.ask.conversations.ensure";
    AlexaConversations["EnsureArgOrder"] = "com.amazon.alexa.ask.conversations.EnsureArgOrder";
    AlexaConversations["Enumeration"] = "com.amazon.alexa.ask.conversations.Enumeration";
    AlexaConversations["Eq"] = "com.amazon.alexa.ask.conversations.equal";
    AlexaConversations["events"] = "com.amazon.alexa.ask.conversations.events";
    AlexaConversations["exists"] = "com.amazon.alexa.ask.conversations.exists";
    AlexaConversations["expect"] = "com.amazon.alexa.ask.conversations.expect";
    AlexaConversations["extendCatalog"] = "com.amazon.alexa.ask.conversations.extendCatalog";
    AlexaConversations["generated"] = "com.amazon.alexa.ask.conversations.generated";
    AlexaConversations["getAllNames"] = "com.amazon.alexa.ask.conversations.getAllNames";
    AlexaConversations["getListItem"] = "com.amazon.alexa.ask.conversations.getListItem";
    AlexaConversations["Gt"] = "com.amazon.alexa.ask.conversations.greater";
    AlexaConversations["Gte"] = "com.amazon.alexa.ask.conversations.greaterEqual";
    AlexaConversations["Inform"] = "com.amazon.alexa.ask.conversations.Inform";
    AlexaConversations["intent"] = "com.amazon.alexa.ask.conversations.intent";
    AlexaConversations["IntentEvent"] = "com.amazon.alexa.ask.conversations.IntentEvent";
    AlexaConversations["IntentRequest"] = "com.amazon.alexa.ask.conversations.IntentRequest";
    AlexaConversations["InvocationName"] = "com.amazon.alexa.ask.conversations.InvocationName";
    AlexaConversations["Invoke"] = "com.amazon.alexa.ask.conversations.Invoke";
    AlexaConversations["in"] = "com.amazon.alexa.ask.conversations.in";
    AlexaConversations["is"] = "com.amazon.alexa.ask.conversations.is";
    AlexaConversations["isInCatalog"] = "com.amazon.alexa.ask.conversations.isInCatalog";
    AlexaConversations["length"] = "com.amazon.alexa.ask.conversations.length";
    AlexaConversations["listBasedCatalog"] = "com.amazon.alexa.ask.conversations.listBasedCatalog";
    AlexaConversations["ListBasedCatalog"] = "com.amazon.alexa.ask.conversations.ListBasedCatalog";
    AlexaConversations["locale"] = "com.amazon.alexa.ask.conversations.locale";
    AlexaConversations["Locale"] = "com.amazon.alexa.ask.conversations.Locale";
    AlexaConversations["Lt"] = "com.amazon.alexa.ask.conversations.less";
    AlexaConversations["Lte"] = "com.amazon.alexa.ask.conversations.lessEqual";
    AlexaConversations["MultiModalResponse"] = "com.amazon.alexa.ask.conversations.MultiModalResponse";
    AlexaConversations["MultiModalEvent"] = "com.amazon.alexa.ask.conversations.MultiModalEvent";
    AlexaConversations["multiModalEvent"] = "com.amazon.alexa.ask.conversations.multiModalEvent";
    AlexaConversations["Neq"] = "com.amazon.alexa.ask.conversations.notEqual";
    AlexaConversations["Not"] = "com.amazon.alexa.ask.conversations.not";
    AlexaConversations["Notify"] = "com.amazon.alexa.ask.conversations.Notify";
    AlexaConversations["Offer"] = "com.amazon.alexa.ask.conversations.Offer";
    AlexaConversations["Or"] = "com.amazon.alexa.ask.conversations.or";
    AlexaConversations["Path"] = "com.amazon.alexa.ask.conversations.Path";
    AlexaConversations["received"] = "com.amazon.alexa.ask.conversations.received";
    AlexaConversations["referable"] = "com.amazon.alexa.ask.conversations.referable";
    AlexaConversations["ReqAlt"] = "com.amazon.alexa.ask.conversations.ReqAlt";
    AlexaConversations["ReqMore"] = "com.amazon.alexa.ask.conversations.ReqMore";
    AlexaConversations["Request"] = "com.amazon.alexa.ask.conversations.Request";
    AlexaConversations["RequestArguments"] = "com.amazon.alexa.ask.conversations.RequestArguments";
    AlexaConversations["ResetAllArguments"] = "com.amazon.alexa.ask.conversations.ResetAllArguments";
    AlexaConversations["ResetArguments"] = "com.amazon.alexa.ask.conversations.ResetArguments";
    AlexaConversations["response"] = "com.amazon.alexa.ask.conversations.response";
    AlexaConversations["ResponseAct"] = "com.amazon.alexa.ask.conversations.ResponseAct";
    AlexaConversations["RequestAct"] = "com.amazon.alexa.ask.conversations.RequestAct";
    AlexaConversations["size"] = "com.amazon.alexa.ask.conversations.size";
    AlexaConversations["skill"] = "com.amazon.alexa.ask.conversations.skill";
    AlexaConversations["Skill"] = "com.amazon.alexa.ask.conversations.Skill";
    AlexaConversations["SkillLevelResponses"] = "com.amazon.alexa.ask.conversations.SkillLevelResponses";
    AlexaConversations["surfaceForm"] = "com.amazon.alexa.ask.conversations.surfaceForm";
    AlexaConversations["taskContextProjection"] = "com.amazon.alexa.ask.conversations.taskContextProjection";
    AlexaConversations["TaskContextProjection"] = "com.amazon.alexa.ask.conversations.TaskContextProjection";
    AlexaConversations["TaskContextProjectionInfo"] = "com.amazon.alexa.ask.conversations.TaskContextProjectionInfo";
    AlexaConversations["TaskContextProjectionType"] = "com.amazon.alexa.ask.conversations.TaskContextProjectionType";
    AlexaConversations["touch"] = "com.amazon.alexa.ask.conversations.touch";
    AlexaConversations["Type"] = "com.amazon.alexa.ask.conversations.Type";
    AlexaConversations["userSays"] = "com.amazon.alexa.ask.conversations.userSays";
    AlexaConversations["Utterance"] = "com.amazon.alexa.ask.conversations.Utterance";
    AlexaConversations["UtteranceEvent"] = "com.amazon.alexa.ask.conversations.UtteranceEvent";
    AlexaConversations["utterances"] = "com.amazon.alexa.ask.conversations.utterances";
    AlexaConversations["variations"] = "com.amazon.alexa.ask.conversations.variations";
    AlexaConversations["welcome"] = "com.amazon.alexa.ask.conversations.welcome";
    AlexaConversations["when"] = "com.amazon.alexa.ask.conversations.when";
    AlexaConversations["validateArg"] = "com.amazon.alexa.ask.conversations.validateArg";
})(AlexaConversations = exports.AlexaConversations || (exports.AlexaConversations = {}));
function ActionTypeName(n) {
    return `${AlexaConversations.Action}${n}`;
}
exports.ActionTypeName = ActionTypeName;
function DialogTypeName(n) {
    return `${AlexaConversations.Dialog}${n}`;
}
exports.DialogTypeName = DialogTypeName;
function FunctionTypeName(n) {
    return `${AlexaSchema.Function}${n}`;
}
exports.FunctionTypeName = FunctionTypeName;
exports.OPERATOR_ACTIONS_SET = new Set([
    // Logical operations
    AlexaConversations.Eq,
    AlexaConversations.Gt,
    AlexaConversations.Gte,
    AlexaConversations.Lt,
    AlexaConversations.Lte,
    AlexaConversations.Neq,
    AlexaConversations.Not,
    AlexaConversations.And,
    AlexaConversations.Or,
    AlexaConversations.is,
]);
var AnnotationTarget;
(function (AnnotationTarget) {
    /**
     * Annotation is allowed on an `action` declaration.
     */
    AnnotationTarget["Action"] = "Action";
    /**
     * Annotation is allowed on a `argument declaration` expression.
     */
    AnnotationTarget["Argument"] = "Argument";
    /**
     * Annotation is allowed on a `call` expression.
     */
    AnnotationTarget["Call"] = "Call";
    /**
     * Annotation is allowed on a `condition` expression.
     */
    AnnotationTarget["Condition"] = "Condition";
    /**
     * Annotation is allowed on a `dialog` declaration.
     */
    AnnotationTarget["Dialog"] = "Dialog";
    /**
     * Annotation is allowed on a `enum` declaration.
     */
    AnnotationTarget["Enum"] = "Enum";
    /**
     * Annotation is allowed on a `enum` declaration's item.
     */
    AnnotationTarget["EnumItem"] = "EnumItem";
    /**
     * Annotation is allowed on a `interaction` declaration.
     */
    AnnotationTarget["Interaction"] = "Interaction";
    /**
     * Annotation is allowed on a `property` declaration.
     */
    AnnotationTarget["Property"] = "Property";
    /**
     * Annotation is allowed on a `property reference` expression.
     */
    AnnotationTarget["PropertyReference"] = "PropertyReference";
    /**
     * Annotation is allowed on a `name` declaration.
     */
    AnnotationTarget["NameDeclaration"] = "NameDeclaration";
    /**
     * Annotation is allowed on a `name reference` expression.
     */
    AnnotationTarget["NameReference"] = "NameReference";
    /**
     * Annotation is allowed on a `sample` block.
     */
    AnnotationTarget["Sample"] = "Sample";
    /**
     * Annotation is allowed on a `type` declaration.
     */
    AnnotationTarget["Type"] = "Type";
    /**
     * Annotation is allowed on a `type reference` expression.
     */
    AnnotationTarget["TypeReference"] = "TypeReference";
})(AnnotationTarget = exports.AnnotationTarget || (exports.AnnotationTarget = {}));
(function (AnnotationTarget) {
    function from(kind) {
        return fromKindMapping[kind];
    }
    AnnotationTarget.from = from;
})(AnnotationTarget = exports.AnnotationTarget || (exports.AnnotationTarget = {}));
const fromKindMapping = {
    ActionDeclaration: AnnotationTarget.Action,
    ArgumentDeclaration: AnnotationTarget.Argument,
    Call: AnnotationTarget.Call,
    DialogDeclaration: AnnotationTarget.Dialog,
    EnumDeclaration: AnnotationTarget.Enum,
    Condition: AnnotationTarget.Condition,
    EnumItemDeclaration: AnnotationTarget.EnumItem,
    InteractionDeclaration: AnnotationTarget.Interaction,
    NameDeclaration: AnnotationTarget.NameDeclaration,
    NameReference: AnnotationTarget.NameReference,
    PropertyReference: AnnotationTarget.PropertyReference,
    Sample: AnnotationTarget.Sample,
    TypeDeclaration: AnnotationTarget.Type,
    TypeProperty: AnnotationTarget.Property,
    TypeReference: AnnotationTarget.TypeReference,
};
const annotationTargets = new Set(Object.values(fromKindMapping));
function isAnnotationTarget(a) {
    return annotationTargets.has(a);
}
exports.isAnnotationTarget = isAnnotationTarget;
var AlexaConversationPrompts;
(function (AlexaConversationPrompts) {
    AlexaConversationPrompts["Bye"] = "AlexaConversationsBye";
    AlexaConversationPrompts["OutOfDomain"] = "AlexaConversationsOutOfDomain";
    AlexaConversationPrompts["ProvideHelp"] = "AlexaConversationsProvideHelp";
    AlexaConversationPrompts["RequestMore"] = "AlexaConversationsRequestMore";
    AlexaConversationPrompts["Welcome"] = "AlexaConversationsWelcome";
})(AlexaConversationPrompts = exports.AlexaConversationPrompts || (exports.AlexaConversationPrompts = {}));
exports.BuiltinsNamespace = "com.amazon.ask.types.builtins.AMAZON";
var BuiltIns;
(function (BuiltIns) {
    BuiltIns["NUMBER"] = "com.amazon.ask.types.builtins.AMAZON.NUMBER";
    BuiltIns["FOUR_DIGIT_NUMBER"] = "com.amazon.ask.types.builtins.AMAZON.FOUR_DIGIT_NUMBER";
    BuiltIns["Ordinal"] = "com.amazon.ask.types.builtins.AMAZON.Ordinal";
    BuiltIns["Actor"] = "com.amazon.ask.types.builtins.AMAZON.Actor";
    BuiltIns["AdministrativeArea"] = "com.amazon.ask.types.builtins.AMAZON.AdministrativeArea";
    BuiltIns["AggregateRating"] = "com.amazon.ask.types.builtins.AMAZON.AggregateRating";
    BuiltIns["Airline"] = "com.amazon.ask.types.builtins.AMAZON.Airline";
    BuiltIns["Airport"] = "com.amazon.ask.types.builtins.AMAZON.Airport";
    BuiltIns["Anaphor"] = "com.amazon.ask.types.builtins.AMAZON.Anaphor";
    BuiltIns["Animal"] = "com.amazon.ask.types.builtins.AMAZON.Animal";
    BuiltIns["Artist"] = "com.amazon.ask.types.builtins.AMAZON.Artist";
    BuiltIns["AT_CITY"] = "com.amazon.ask.types.builtins.AMAZON.AT_CITY";
    BuiltIns["AT_REGION"] = "com.amazon.ask.types.builtins.AMAZON.AT_REGION";
    BuiltIns["Athlete"] = "com.amazon.ask.types.builtins.AMAZON.Athlete";
    BuiltIns["Author"] = "com.amazon.ask.types.builtins.AMAZON.Author";
    BuiltIns["Book"] = "com.amazon.ask.types.builtins.AMAZON.Book";
    BuiltIns["BookSeries"] = "com.amazon.ask.types.builtins.AMAZON.BookSeries";
    BuiltIns["BroadcastChannel"] = "com.amazon.ask.types.builtins.AMAZON.BroadcastChannel";
    BuiltIns["City"] = "com.amazon.ask.types.builtins.AMAZON.City";
    BuiltIns["CivicStructure"] = "com.amazon.ask.types.builtins.AMAZON.CivicStructure";
    BuiltIns["Color"] = "com.amazon.ask.types.builtins.AMAZON.Color";
    BuiltIns["Comic"] = "com.amazon.ask.types.builtins.AMAZON.Comic";
    BuiltIns["Corporation"] = "com.amazon.ask.types.builtins.AMAZON.Corporation";
    BuiltIns["Country"] = "com.amazon.ask.types.builtins.AMAZON.Country";
    BuiltIns["CreativeWorkType"] = "com.amazon.ask.types.builtins.AMAZON.CreativeWorkType";
    BuiltIns["DATE"] = "com.amazon.ask.types.builtins.AMAZON.DATE";
    BuiltIns["DayOfWeek"] = "com.amazon.ask.types.builtins.AMAZON.DayOfWeek";
    BuiltIns["DE_CITY"] = "com.amazon.ask.types.builtins.AMAZON.DE_CITY";
    BuiltIns["DE_FIRST_NAME"] = "com.amazon.ask.types.builtins.AMAZON.DE_FIRST_NAME";
    BuiltIns["DE_REGION"] = "com.amazon.ask.types.builtins.AMAZON.DE_REGION";
    BuiltIns["Decimal"] = "com.amazon.ask.types.builtins.AMAZON.Decimal";
    BuiltIns["Dessert"] = "com.amazon.ask.types.builtins.AMAZON.Dessert";
    BuiltIns["DeviceType"] = "com.amazon.ask.types.builtins.AMAZON.DeviceType";
    BuiltIns["Director"] = "com.amazon.ask.types.builtins.AMAZON.Director";
    BuiltIns["Drink"] = "com.amazon.ask.types.builtins.AMAZON.Drink";
    BuiltIns["DURATION"] = "com.amazon.ask.types.builtins.AMAZON.DURATION";
    BuiltIns["EducationalOrganization"] = "com.amazon.ask.types.builtins.AMAZON.EducationalOrganization";
    BuiltIns["EUROPE_CITY"] = "com.amazon.ask.types.builtins.AMAZON.EUROPE_CITY";
    BuiltIns["EventType"] = "com.amazon.ask.types.builtins.AMAZON.EventType";
    BuiltIns["Festival"] = "com.amazon.ask.types.builtins.AMAZON.Festival";
    BuiltIns["FictionalCharacter"] = "com.amazon.ask.types.builtins.AMAZON.FictionalCharacter";
    BuiltIns["FinancialService"] = "com.amazon.ask.types.builtins.AMAZON.FinancialService";
    BuiltIns["FirstName"] = "com.amazon.ask.types.builtins.AMAZON.FirstName";
    BuiltIns["Food"] = "com.amazon.ask.types.builtins.AMAZON.Food";
    BuiltIns["FoodEstablishment"] = "com.amazon.ask.types.builtins.AMAZON.FoodEstablishment";
    BuiltIns["Game"] = "com.amazon.ask.types.builtins.AMAZON.Game";
    BuiltIns["GB_CITY"] = "com.amazon.ask.types.builtins.AMAZON.GB_CITY";
    BuiltIns["GB_FIRST_NAME"] = "com.amazon.ask.types.builtins.AMAZON.GB_FIRST_NAME";
    BuiltIns["GB_REGION"] = "com.amazon.ask.types.builtins.AMAZON.GB_REGION";
    BuiltIns["Genre"] = "com.amazon.ask.types.builtins.AMAZON.Genre";
    BuiltIns["Landform"] = "com.amazon.ask.types.builtins.AMAZON.Landform";
    BuiltIns["LandmarksOrHistoricalBuildings"] = "com.amazon.ask.types.builtins.AMAZON.LandmarksOrHistoricalBuildings";
    BuiltIns["Language"] = "com.amazon.ask.types.builtins.AMAZON.Language";
    BuiltIns["Literal"] = "com.amazon.ask.types.builtins.AMAZON.Literal";
    BuiltIns["LocalBusiness"] = "com.amazon.ask.types.builtins.AMAZON.LocalBusiness";
    BuiltIns["LocalBusinessType"] = "com.amazon.ask.types.builtins.AMAZON.LocalBusinessType";
    BuiltIns["MedicalOrganization"] = "com.amazon.ask.types.builtins.AMAZON.MedicalOrganization";
    BuiltIns["Month"] = "com.amazon.ask.types.builtins.AMAZON.Month";
    BuiltIns["Movie"] = "com.amazon.ask.types.builtins.AMAZON.Movie";
    BuiltIns["MovieTheater"] = "com.amazon.ask.types.builtins.AMAZON.MovieTheater";
    BuiltIns["MusicAlbum"] = "com.amazon.ask.types.builtins.AMAZON.MusicAlbum";
    BuiltIns["MusicCreativeWorkType"] = "com.amazon.ask.types.builtins.AMAZON.MusicCreativeWorkType";
    BuiltIns["MusicEvent"] = "com.amazon.ask.types.builtins.AMAZON.MusicEvent";
    BuiltIns["MusicGroup"] = "com.amazon.ask.types.builtins.AMAZON.MusicGroup";
    BuiltIns["Musician"] = "com.amazon.ask.types.builtins.AMAZON.Musician";
    BuiltIns["MusicPlaylist"] = "com.amazon.ask.types.builtins.AMAZON.MusicPlaylist";
    BuiltIns["MusicRecording"] = "com.amazon.ask.types.builtins.AMAZON.MusicRecording";
    BuiltIns["MusicVenue"] = "com.amazon.ask.types.builtins.AMAZON.MusicVenue";
    BuiltIns["MusicVideo"] = "com.amazon.ask.types.builtins.AMAZON.MusicVideo";
    BuiltIns["Organization"] = "com.amazon.ask.types.builtins.AMAZON.Organization";
    BuiltIns["PartySize"] = "com.amazon.ask.types.builtins.AMAZON.PartySize";
    BuiltIns["Person"] = "com.amazon.ask.types.builtins.AMAZON.Person";
    BuiltIns["PhoneNumber"] = "com.amazon.ask.types.builtins.AMAZON.PhoneNumber";
    BuiltIns["PostalAddress"] = "com.amazon.ask.types.builtins.AMAZON.PostalAddress";
    BuiltIns["Professional"] = "com.amazon.ask.types.builtins.AMAZON.Professional";
    BuiltIns["ProfessionalType"] = "com.amazon.ask.types.builtins.AMAZON.ProfessionalType";
    BuiltIns["RadioChannel"] = "com.amazon.ask.types.builtins.AMAZON.RadioChannel";
    BuiltIns["Region"] = "com.amazon.ask.types.builtins.AMAZON.Region";
    BuiltIns["RelativePosition"] = "com.amazon.ask.types.builtins.AMAZON.RelativePosition";
    BuiltIns["Residence"] = "com.amazon.ask.types.builtins.AMAZON.Residence";
    BuiltIns["Room"] = "com.amazon.ask.types.builtins.AMAZON.Room";
    BuiltIns["ScreeningEvent"] = "com.amazon.ask.types.builtins.AMAZON.ScreeningEvent";
    BuiltIns["Service"] = "com.amazon.ask.types.builtins.AMAZON.Service";
    BuiltIns["SocialMediaPlatform"] = "com.amazon.ask.types.builtins.AMAZON.SocialMediaPlatform";
    BuiltIns["SoftwareApplication"] = "com.amazon.ask.types.builtins.AMAZON.SoftwareApplication";
    BuiltIns["SoftwareGame"] = "com.amazon.ask.types.builtins.AMAZON.SoftwareGame";
    BuiltIns["Sport"] = "com.amazon.ask.types.builtins.AMAZON.Sport";
    BuiltIns["SportsEvent"] = "com.amazon.ask.types.builtins.AMAZON.SportsEvent";
    BuiltIns["SportsTeam"] = "com.amazon.ask.types.builtins.AMAZON.SportsTeam";
    BuiltIns["StreetAddress"] = "com.amazon.ask.types.builtins.AMAZON.StreetAddress";
    BuiltIns["StreetName"] = "com.amazon.ask.types.builtins.AMAZON.StreetName";
    BuiltIns["TelevisionChannel"] = "com.amazon.ask.types.builtins.AMAZON.TelevisionChannel";
    BuiltIns["TIME"] = "com.amazon.ask.types.builtins.AMAZON.TIME";
    BuiltIns["TVEpisode"] = "com.amazon.ask.types.builtins.AMAZON.TVEpisode";
    BuiltIns["TVSeason"] = "com.amazon.ask.types.builtins.AMAZON.TVSeason";
    BuiltIns["US_CITY"] = "com.amazon.ask.types.builtins.AMAZON.US_CITY";
    BuiltIns["US_FIRST_NAME"] = "com.amazon.ask.types.builtins.AMAZON.US_FIRST_NAME";
    BuiltIns["US_STATE"] = "com.amazon.ask.types.builtins.AMAZON.US_STATE";
    BuiltIns["VideoGame"] = "com.amazon.ask.types.builtins.AMAZON.VideoGame";
    BuiltIns["VisualModeTrigger"] = "com.amazon.ask.types.builtins.AMAZON.VisualModeTrigger";
    BuiltIns["WeatherCondition"] = "com.amazon.ask.types.builtins.AMAZON.WeatherCondition";
    BuiltIns["WrittenCreativeWorkType"] = "com.amazon.ask.types.builtins.AMAZON.WrittenCreativeWorkType";
})(BuiltIns = exports.BuiltIns || (exports.BuiltIns = {}));
/**
 * Represents the `com.amazon.alexa.ask.conversations.Locale` enum.
 * @see https://developer.amazon.com/en-US/docs/alexa/custom-skills/develop-skills-in-multiple-languages.html
 */
var Locale;
(function (Locale) {
    Locale["en_US"] = "en_US";
    Locale["en_GB"] = "en_GB";
    Locale["en_AU"] = "en_AU";
    Locale["en_CA"] = "en_CA";
    Locale["en_IN"] = "en_IN";
    Locale["de_DE"] = "de_DE";
    Locale["es_US"] = "es_US";
    Locale["ja_JP"] = "ja_JP";
    Locale["ar_SA"] = "ar_SA";
    Locale["es_ES"] = "es_ES";
    Locale["es_MX"] = "es_MX";
    Locale["fr_CA"] = "fr_CA";
    Locale["hi_IN"] = "hi_IN";
    Locale["it_IT"] = "it_IT";
    Locale["pt_BR"] = "pt_BR";
    Locale["fr_FR"] = "fr_FR";
})(Locale = exports.Locale || (exports.Locale = {}));
(function (Locale) {
    function toUnderscoreLocale(locale) {
        return locale.replace("-", "_");
    }
    Locale.toUnderscoreLocale = toUnderscoreLocale;
})(Locale = exports.Locale || (exports.Locale = {}));
