// @alexa/acdl/dist/cjs/ast/property-reference.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.PropertyReference = exports.isPropRef = void 0;
const guard_1 = require("./guard");
const tree_1 = require("./tree");
/**
 * Asserts if the Expression is a `PropertyReference`.
 */
exports.isPropRef = (0, guard_1.guard)("PropertyReference");
/**
 * Represents the access of a Property.
 */
class PropertyReference extends tree_1.Tree {
    constructor(
    /**
     * Name of the property being accessed.
     */
    name, 
    /**
     * Expression containing the property being accessed.
     *
     * Names, Calls and other PropertyAccess nodes are supported.
     */
    expression, annotations, loc) {
        super("PropertyReference", loc);
        Object.defineProperty(this, "name", {
            enumerable: true,
            configurable: true,
            writable: true,
            value: name
        });
        Object.defineProperty(this, "expression", {
            enumerable: true,
            configurable: true,
            writable: true,
            value: expression
        });
        Object.defineProperty(this, "annotations", {
            enumerable: true,
            configurable: true,
            writable: true,
            value: annotations
        });
        this.setAsParentOn(name);
        this.setAsParentOn(expression);
        this.setAsParentOn(annotations);
    }
    clone() {
        var _a, _b, _c;
        return new PropertyReference((_a = this.name) === null || _a === void 0 ? void 0 : _a.clone(), (_b = this.expression) === null || _b === void 0 ? void 0 : _b.clone(), (_c = this.annotations) === null || _c === void 0 ? void 0 : _c.map((a) => a.clone()), this.loc);
    }
    /**
     * Print the full prop chain with dot as the separator.
     *
     * @returns the dot-joined string from the called node to its root
     */
    printPropertyChainThroughRoot(sep) {
        var _a, _b, _c;
        const chainingProps = [(_a = this.name) === null || _a === void 0 ? void 0 : _a.name];
        let temp = this.expression;
        while ((temp === null || temp === void 0 ? void 0 : temp.kind) === "PropertyReference" && ((_b = temp.name) === null || _b === void 0 ? void 0 : _b.name)) {
            chainingProps.push(temp.name.name);
            temp = temp.expression;
        }
        if (temp && ((_c = temp.name) === null || _c === void 0 ? void 0 : _c.name)) {
            chainingProps.push(temp.name.name);
        }
        return chainingProps.reverse().join(sep || ".");
    }
}
exports.PropertyReference = PropertyReference;
(function (PropertyReference) {
    /**
     * Tries to resolves a PropertyReference to a QualifiedName.
     *
     * When we parse qualified names, e.g. `com.amazon.MyIntent`, it is ambiguous
     * as to whether this is
     * 1. `PropertyReference(MyIntent, PropertyReference(amazon, NameReference(com)))`
     * 2. `NameReference(com.amazon.MyIntent)`
     *
     * This function flattens a PropertyReference tree to a qualified name so that it can be checked.
     *
     * @param expr
     */
    function tryResolveQualifiedName(expr) {
        var _a, _b;
        if ((expr === null || expr === void 0 ? void 0 : expr.kind) === "Call") {
            return undefined;
        }
        if ((expr === null || expr === void 0 ? void 0 : expr.kind) === "NameReference") {
            return (_a = expr.name) === null || _a === void 0 ? void 0 : _a.name;
        }
        if ((expr === null || expr === void 0 ? void 0 : expr.kind) === "PropertyReference") {
            const prevName = tryResolveQualifiedName(expr.expression);
            if (prevName) {
                return `${prevName}.${(_b = expr.name) === null || _b === void 0 ? void 0 : _b.name}`;
            }
        }
        return undefined;
    }
    PropertyReference.tryResolveQualifiedName = tryResolveQualifiedName;
    /**
     * Tries to resolves a PropertyReference tree to a SourceLocation spanning the entire tree as
     * if it were a qualified name.
     *
     * When we parse qualified names, e.g. `com.amazon.MyIntent`, it is ambiguous
     * as to whether this is
     * 1. `PropertyReference(MyIntent, PropertyReference(amazon, NameReference(com)))`
     * 2. `NameReference(com.amazon.MyIntent)`
     *
     * @param expr
     */
    function tryResolveQualifiedNameLocation(expr) {
        var _a;
        if (expr === undefined) {
            return undefined;
        }
        const end = (_a = expr === null || expr === void 0 ? void 0 : expr.loc) === null || _a === void 0 ? void 0 : _a.end;
        const begin = findBeginning(expr);
        if (end && begin) {
            return {
                begin,
                end,
            };
        }
        return undefined;
        function findBeginning(expr) {
            var _a;
            if ((expr === null || expr === void 0 ? void 0 : expr.kind) === "NameReference") {
                return (_a = expr.loc) === null || _a === void 0 ? void 0 : _a.begin;
            }
            if ((expr === null || expr === void 0 ? void 0 : expr.kind) === "PropertyReference") {
                return findBeginning(expr.expression);
            }
            return undefined;
        }
    }
    PropertyReference.tryResolveQualifiedNameLocation = tryResolveQualifiedNameLocation;
})(PropertyReference = exports.PropertyReference || (exports.PropertyReference = {}));
