// @alexa/acdl/dist/cjs/ast/source-location.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.isBefore = exports.isAfter = exports.locationContains = exports.betweenLocations = exports.appendLocation = void 0;
function appendLocation(location, num) {
    return {
        ...location,
        end: typeof num === "object"
            ? num.end
            : {
                ...location.end,
                character: location.end.character + num,
            },
    };
}
exports.appendLocation = appendLocation;
function betweenLocations(left, right) {
    return {
        begin: left.begin,
        end: right.end,
    };
}
exports.betweenLocations = betweenLocations;
function locationContains(location, position, inclusive = false) {
    return isAfter(location.begin, position) && isBefore(location.end, position, inclusive);
}
exports.locationContains = locationContains;
/**
 * Returns true if `b` is after `a`.
 */
function isAfter(a, b) {
    return b.line >= a.line && (b.line > a.line || b.character >= a.character);
}
exports.isAfter = isAfter;
/**
 * Returns true if `b` is before `a`.
 */
function isBefore(a, b, inclusive = false) {
    return b.line <= a.line && (b.line < a.line || b.character <= (inclusive ? a.character + 1 : a.character));
}
exports.isBefore = isBefore;
