// @alexa/acdl/dist/cjs/ast/tree.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.Tree = void 0;
class Tree {
    constructor(kind, loc) {
        Object.defineProperty(this, "kind", {
            enumerable: true,
            configurable: true,
            writable: true,
            value: kind
        });
        Object.defineProperty(this, "loc", {
            enumerable: true,
            configurable: true,
            writable: true,
            value: loc
        });
        /**
         * Surrounding Context in which this node is found.
         */
        Object.defineProperty(this, "context", {
            enumerable: true,
            configurable: true,
            writable: true,
            value: void 0
        });
        /**
         * Pointer to the Node that came before this one in a block sequence.
         *
         * This is only set for nodes in a Scope.
         */
        Object.defineProperty(this, "previous", {
            enumerable: true,
            configurable: true,
            writable: true,
            value: void 0
        });
        /**
         * Pointer to the Node that came after this one in a block sequence.
         *
         * This is only set for nodes in a Scope.
         */
        Object.defineProperty(this, "next", {
            enumerable: true,
            configurable: true,
            writable: true,
            value: void 0
        });
    }
    /**
     * The URI this tree node is contained within.
     */
    get uri() {
        var _a;
        return (_a = this.getModule()) === null || _a === void 0 ? void 0 : _a.uri;
    }
    /**
     * Pretty Print this Node.
     *
     * @param level indentation level
     */
    toString(level = 0) {
        return `${this.kind}(\n${indent(level)}${Object.entries(this)
            .map(([name, value]) => {
            if (name === "previous") {
            }
            else if (name === "context") {
                return `${indent(level)}context = ${this.printHierarchy().join("->")}`;
            }
            else if (name === "loc") {
            }
            else if (Array.isArray(value)) {
                return `${indent(level)}${name} = [\n${indent(level)}${value
                    .map((v) => indent(level + 1) + v.toString(level + 2))
                    .join(`,\n${indent(level)}`)}\n${indent(level)}]`;
            }
            else if (typeof value !== "function") {
                return `${indent(level)}${name} = ${typeof value === "object" ? value.toString(level + 1) : typeof value === "string" ? `"${value}"` : "null"}`;
            }
            return undefined;
        })
            .filter((s) => s !== undefined)
            .join(`,\n${indent(level)}`)})`;
    }
    printHierarchy() {
        var _a;
        return [this.kind, ...(((_a = this.context) === null || _a === void 0 ? void 0 : _a.printHierarchy()) || [])];
    }
    setAsParentOn(t) {
        if (t !== undefined) {
            if (Array.isArray(t)) {
                for (const i of t) {
                    if (i !== undefined) {
                        i.context = this;
                    }
                }
            }
            else {
                t.context = this;
            }
        }
    }
    findInContext(match) {
        let curr = this;
        while (curr !== undefined) {
            if (match(curr)) {
                return curr;
            }
            curr = curr.context;
        }
        return undefined;
    }
    /**
     * Find the Sample an expression came from.
     *
     * @param scope expression scope to find the containing module of
     */
    getDialog() {
        return this.findInContext((n) => (n === null || n === void 0 ? void 0 : n.kind) === "DialogDeclaration");
    }
    /**
     * Find the Sample an expression came from.
     */
    getSample() {
        return this.findInContext((n) => (n === null || n === void 0 ? void 0 : n.kind) === "Sample");
    }
    /**
     * Find the Module or ACDL file an expression came from.
     */
    getModule() {
        return this.findInContext((n) => (n === null || n === void 0 ? void 0 : n.kind) === "Module");
    }
    /**
     * Find the nearest Correction in this Node's context.
     */
    getCorrection() {
        return this.findInContext((n) => (n === null || n === void 0 ? void 0 : n.kind) === "Correction");
    }
}
exports.Tree = Tree;
function indent(level) {
    return " ".repeat(level);
}
