// @alexa/acdl/dist/esm/ast/response-template.js
import { Tree } from "./tree";
/**
 * Types of response templates supported by the compiler.
 */
export var ResponseTemplateType;
(function (ResponseTemplateType) {
    /**
     * APL display template
     */
    ResponseTemplateType["Display"] = "display";
    /**
     * APL-A prompt template.
     */
    ResponseTemplateType["Prompt"] = "prompt";
})(ResponseTemplateType || (ResponseTemplateType = {}));
(function (ResponseTemplateType) {
    ResponseTemplateType.All = [
        ResponseTemplateType.Display,
        ResponseTemplateType.Prompt,
    ];
})(ResponseTemplateType || (ResponseTemplateType = {}));
/**
 * Node type to represent an externally loaded response template.
 *
 * This node will be removed in Beta 2 when we have generated prompts and the
 * ability to require them from disk.
 */
export class ResponseTemplate extends Tree {
    constructor(
    /**
     * URI of the folder containing the APLA `document.json` file.
     */
    templateUri, 
    /**
     * Name of the prompt.
     *
     * Example:
     * ```
     * // name: `hello`
     * response/prompts/hello/document.json
     * ```
     */
    name, type, loc) {
        super("ResponseTemplate", loc);
        Object.defineProperty(this, "templateUri", {
            enumerable: true,
            configurable: true,
            writable: true,
            value: templateUri
        });
        Object.defineProperty(this, "name", {
            enumerable: true,
            configurable: true,
            writable: true,
            value: name
        });
        Object.defineProperty(this, "type", {
            enumerable: true,
            configurable: true,
            writable: true,
            value: type
        });
    }
    get uri() {
        return this.templateUri;
    }
    /**
     * Check if this is an APL Display Response Template.
     */
    isDisplay() {
        return this.type === ResponseTemplateType.Display;
    }
    /**
     * Check if this is an APLA Prompt Response Template.
     */
    isPrompt() {
        return this.type === ResponseTemplateType.Prompt;
    }
    clone() {
        return new ResponseTemplate(this.templateUri, this.name, this.type, this.loc);
    }
}
