// @alexa/acdl/dist/cjs/ask-cli-config.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.AskCliConfig = void 0;
const os = __importStar(require("os"));
const fs = __importStar(require("fs-extra"));
const uuid_1 = require("uuid");
class AskCliConfig {
    /**
     * Load the user's ASK CLI Config.
     *
     * @param path path of the cli_config - defaults to `~/.ask/cli_config`
     */
    constructor() {
        Object.defineProperty(this, "askCliConfig", {
            enumerable: true,
            configurable: true,
            writable: true,
            value: void 0
        });
        Object.defineProperty(this, "cliConfigPath", {
            enumerable: true,
            configurable: true,
            writable: true,
            value: void 0
        });
        Object.defineProperty(this, "errorLoadingAskCliConfig", {
            enumerable: true,
            configurable: true,
            writable: true,
            value: void 0
        });
        this.cliConfigPath = os.homedir() + "/.ask/cli_config";
        let cliConfigString;
        try {
            cliConfigString = fs.readFileSync(this.cliConfigPath).toString("utf8");
        }
        catch (err) {
            this.errorLoadingAskCliConfig = `cannot read config file: ${this.cliConfigPath}. Run 'ask configure'.`;
            return;
        }
        try {
            this.askCliConfig = JSON.parse(cliConfigString);
        }
        catch (err) {
            this.errorLoadingAskCliConfig = `failed to parse config file: ${this.cliConfigPath}`;
            return;
        }
    }
    hasConfig() {
        return !!this.askCliConfig;
    }
    createMachineId() {
        const machine_id = (0, uuid_1.v4)();
        try {
            fs.writeFileSync(this.cliConfigPath, JSON.stringify({ ...(this.askCliConfig || {}), machine_id }, null, 2), "utf-8");
        }
        catch (error) {
            return "";
        }
        return machine_id;
    }
    getLwaProfile(profileName) {
        if (this.errorLoadingAskCliConfig) {
            throw new Error(this.errorLoadingAskCliConfig);
        }
        const lwaProfile = this.askCliConfig.profiles[profileName];
        if (lwaProfile === undefined) {
            this.errorLoadingAskCliConfig = `could not load profile '${profileName}' from ${this.cliConfigPath}`;
        }
        return lwaProfile;
    }
    getShareUsage() {
        const shareUsage = this.askCliConfig.share_usage;
        if (shareUsage !== undefined)
            return !!shareUsage;
        return true;
    }
    getMachineId() {
        return this.askCliConfig.machine_id;
    }
}
exports.AskCliConfig = AskCliConfig;
