// @alexa/acdl/dist/cjs/interaction-model-merger.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.InteractionModelMerger = exports.SlotTypeMerger = exports.SlotTypeValueMerger = exports.IntentSlotMerger = exports.mergeArrays = void 0;
function toMap(items, keySelector) {
    const result = {};
    for (const item of items) {
        result[keySelector(item)] = item;
    }
    return result;
}
function mergeStrings(first, second) {
    return first !== undefined || second !== undefined ? [...new Set([...(first !== null && first !== void 0 ? first : []), ...(second !== null && second !== void 0 ? second : [])]).values()] : undefined;
}
function mergeArrays(first, second, itemMerger) {
    const result = [];
    const firstByName = toMap(first, itemMerger.keySelector);
    const secondByName = toMap(second, itemMerger.keySelector);
    for (const firstItem of first) {
        const secondItem = secondByName[itemMerger.keySelector(firstItem)];
        if (secondItem === undefined) {
            result.push(firstItem);
        }
        else {
            try {
                result.push(itemMerger.merge(firstItem, secondItem));
            }
            catch (e) {
                itemMerger.handleError(firstItem, secondItem, e);
            }
        }
    }
    for (const secondItem of second) {
        const firstItem = firstByName[itemMerger.keySelector(secondItem)];
        if (firstItem === undefined) {
            result.push(secondItem);
        }
    }
    return result;
}
exports.mergeArrays = mergeArrays;
/**
 * Merges intents.
 */
class IntentMerger {
    constructor(intentSlotMerger) {
        Object.defineProperty(this, "intentSlotMerger", {
            enumerable: true,
            configurable: true,
            writable: true,
            value: intentSlotMerger
        });
    }
    keySelector(item) {
        return item.name;
    }
    merge(first, second) {
        var _a, _b;
        const samples = mergeStrings(first.samples, second.samples);
        return {
            ...first,
            slots: mergeArrays((_a = first.slots) !== null && _a !== void 0 ? _a : [], (_b = second.slots) !== null && _b !== void 0 ? _b : [], this.intentSlotMerger),
            ...(samples ? { samples } : {}),
        };
    }
    handleError(first, _second, err) {
        throw new Error(`Failed to merge intent '${first.name}': ${err.message}`);
    }
}
/**
 * Merges intent slots.
 */
class IntentSlotMerger {
    keySelector(item) {
        return item.name;
    }
    merge(first, second) {
        if (first.type === second.type) {
            const samples = mergeStrings(first.samples, second.samples);
            return {
                ...first,
                ...(samples ? { samples } : {}),
            };
        }
        else {
            throw new Error(`Intent slot type differs: '${first.type}' != '${second.type}'`);
        }
    }
    handleError(first, _second, err) {
        throw new Error(`Failed to merge intent slot '${first.name}': ${err.message}`);
    }
}
exports.IntentSlotMerger = IntentSlotMerger;
/**
 * Merges slot type values.
 */
class SlotTypeValueMerger {
    keySelector(item) {
        return item.name.value;
    }
    merge(first, second) {
        var _a;
        if (first.id !== second.id) {
            throw new Error(`Slot type value identifier differs: ${first.id} != ${second.id}`);
        }
        const synonyms = (_a = mergeStrings(first.name.synonyms, second.name.synonyms)) !== null && _a !== void 0 ? _a : [];
        return {
            ...first,
            name: {
                ...first.name,
                ...(synonyms ? { synonyms } : {}),
            },
        };
    }
    handleError(first, _second, err) {
        throw new Error(`Failed to merge slot type value '${first.name.value}': ${err.message}`);
    }
}
exports.SlotTypeValueMerger = SlotTypeValueMerger;
/**
 * Merges slot types.
 */
class SlotTypeMerger {
    constructor(slotTypeValueMerger) {
        Object.defineProperty(this, "slotTypeValueMerger", {
            enumerable: true,
            configurable: true,
            writable: true,
            value: slotTypeValueMerger
        });
    }
    keySelector(item) {
        return item.name;
    }
    merge(first, second) {
        var _a, _b;
        if (!first.valueSupplier && !second.valueSupplier) {
            return {
                ...first,
                values: mergeArrays((_a = first.values) !== null && _a !== void 0 ? _a : [], (_b = second.values) !== null && _b !== void 0 ? _b : [], this.slotTypeValueMerger),
            };
        }
        else {
            throw new Error(`Slot type already exists with different value supplier.`);
        }
    }
    handleError(first, _second, err) {
        throw new Error(`Failed to merge slot type '${first.name}': ${err.message}`);
    }
}
exports.SlotTypeMerger = SlotTypeMerger;
Object.defineProperty(SlotTypeMerger, "INSTANCE", {
    enumerable: true,
    configurable: true,
    writable: true,
    value: new SlotTypeMerger(new SlotTypeValueMerger())
});
/**
 * Merges two interaction models into one (the second one into the first one).
 *
 * Takes all fields from the first interaction model, except for these which are combined together
 * - `intents`
 *   - `slots` - fails if they differ in `type`
 *     - `samples` - union of both samples
 *   - `samples` - union of both samples
 * - `types` - fails if they differ in `valueSupplier`
 *   - `values` - fails if they differ in `id`
 *     - `synonyms` - union of both synonyms
 *
 * ~~~
 * {
 *   interactionModel: {
 *     languageModel: {
 *       intents: [
 *         {
 *           name: 'intentName',
 *           slots: [
 *             {
 *               name: 'slotName',
 *               samples: [...]
 *             }
 *           ]
 *           samples: [...]
 *         }, ...
 *       ]
 *       types: [
 *         {
 *           name: 'typeName'
 *           values: [
 *             {
 *               name: {
 *                 value: 'valueName',
 *                 synonyms: [...]
 *               }
 *             }
 *           ]
 *         }, ...
 *       ]
 *     }
 *   }
 * }
 * ~~~
 *
 * Intents are merged from both models. If same intent is defined in both models, their slots and samples are merged.
 * If the intent slot types do not match, an error will be raised.
 *
 * Slot types are merged from both models. If same slot type is defined in both models, an error will be raised.
 *
 * Note: The code should be written in a way that when new fields are added, they are preserved.
 */
class InteractionModelMerger {
    constructor(intentMerger, slotTypeMerger) {
        Object.defineProperty(this, "intentMerger", {
            enumerable: true,
            configurable: true,
            writable: true,
            value: intentMerger
        });
        Object.defineProperty(this, "slotTypeMerger", {
            enumerable: true,
            configurable: true,
            writable: true,
            value: slotTypeMerger
        });
    }
    /**
     * Merges two interaction models into one (the second one into the first one).
     *
     * @param first First interaction model data.
     * @param second Second interaction model data.
     * @returns The merged interaction model.
     */
    mergeInteractionModels(first, second) {
        return {
            ...first,
            interactionModel: this.mergeInteractionModelSchemas(first.interactionModel, second.interactionModel),
        };
    }
    mergeInteractionModelSchemas(first, second) {
        return {
            ...first,
            languageModel: this.mergeLanguageModels(first.languageModel, second.languageModel),
        };
    }
    mergeLanguageModels(first, second) {
        var _a, _b;
        const types = first.types !== undefined || second.types !== undefined
            ? mergeArrays((_a = first.types) !== null && _a !== void 0 ? _a : [], (_b = second.types) !== null && _b !== void 0 ? _b : [], this.slotTypeMerger)
            : undefined;
        return {
            ...first,
            intents: mergeArrays(first.intents, second.intents, this.intentMerger),
            ...(types ? { types } : {}),
        };
    }
}
exports.InteractionModelMerger = InteractionModelMerger;
Object.defineProperty(InteractionModelMerger, "INSTANCE", {
    enumerable: true,
    configurable: true,
    writable: true,
    value: new InteractionModelMerger(new IntentMerger(new IntentSlotMerger()), SlotTypeMerger.INSTANCE)
});
