// @alexa/acdl/dist/cjs/project-config.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.loadASKIRProject = exports.loadProjectConfig = exports.loadProjectConfigSync = void 0;
const path_1 = __importDefault(require("path"));
const package_1 = require("./package");
const project_1 = require("./project");
const transformUIProject_1 = require("./transformUIProject");
const util_1 = require("./util");
/**
 * Load a project from a directory on disk.
 *
 * We first look for a `package.json` with an `ask` section or an `ask.json` file at the `rootDir` and use
 * it as configuration if it exists.
 *
 * Otherwise we then check if there an `ask-resources.json` file pointing to a `skill-package` folder. Failing that,
 * we then look for a `skill-package` folder at the `rootDir`. In all of these cases, we default to a `1.0.0-beta1`
 * project configuration.
 *
 * @param rootDir path of the project containing `package.json` or `ask.json` configuration file.
 * @param profile ASK CLI profile to use when inspecting `ask-resources.json`.
 */
function loadProjectConfigSync(rootDir = process.cwd(), profile = "default") {
    var _a, _b, _c, _d;
    const pkgJson = (0, package_1.loadPackageConfigSync)(rootDir);
    if (pkgJson !== undefined) {
        return getProjectConfig(rootDir, pkgJson);
    }
    const askResources = (0, util_1.readJsonFileSync)(path_1.default.join(rootDir, "ask-resources.json"));
    const skillPackagePath = path_1.default.join(rootDir, (_d = (_c = (_b = (_a = askResources === null || askResources === void 0 ? void 0 : askResources.profiles) === null || _a === void 0 ? void 0 : _a[profile]) === null || _b === void 0 ? void 0 : _b.skillMetadata) === null || _c === void 0 ? void 0 : _c.src) !== null && _d !== void 0 ? _d : "skill-package");
    if ((0, util_1.existsSync)(skillPackagePath)) {
        return {
            rootDir,
            srcDir: "conversations",
            outDir: "build",
            projectVersion: package_1.ProjectVersion.Beta1,
            sourceRootDir: skillPackagePath,
        };
    }
    throw new Error(`expected 'package.json', 'ask-resources.json' or 'skill-package/' to exist at location: '${rootDir}'`);
}
exports.loadProjectConfigSync = loadProjectConfigSync;
/**
 * Load a project from a directory on disk.
 *
 * We first look for a `package.json` with an `ask` section or an `ask.json` file at the `rootDir` and use
 * it as configuration if it exists.
 *
 * Otherwise we then check if there an `ask-resources.json` file pointing to a `skill-package` folder. Failing that,
 * we then look for a `skill-package` folder at the `rootDir`. In all of these cases, we default to a `1.0.0-beta1`
 * project configuration.
 *
 * @param rootDir path of the project containing `package.json` or `ask.json` configuration file.
 * @param profile ASK CLI profile to use when inspecting `ask-resources.json`.
 */
async function loadProjectConfig(rootDir = process.cwd(), profile = "default") {
    var _a, _b, _c, _d;
    const pkgJson = await (0, package_1.loadPackageConfig)(rootDir);
    if (pkgJson !== undefined) {
        return getProjectConfig(rootDir, pkgJson);
    }
    const askResources = await (0, util_1.readJsonFile)(path_1.default.join(rootDir, "ask-resources.json"));
    const skillPackagePath = path_1.default.join(rootDir, (_d = (_c = (_b = (_a = askResources === null || askResources === void 0 ? void 0 : askResources.profiles) === null || _a === void 0 ? void 0 : _a[profile]) === null || _b === void 0 ? void 0 : _b.skillMetadata) === null || _c === void 0 ? void 0 : _c.src) !== null && _d !== void 0 ? _d : "skill-package");
    if (await (0, util_1.exists)(skillPackagePath)) {
        return {
            rootDir,
            srcDir: "conversations",
            outDir: "build",
            projectVersion: package_1.ProjectVersion.Beta1,
            sourceRootDir: skillPackagePath,
        };
    }
    throw new Error(`expected 'package.json', 'ask-resources.json' or 'skill-package/' to exist at location: '${rootDir}'`);
}
exports.loadProjectConfig = loadProjectConfig;
/**
 * Gets a project corresponding to ASKIR for validation. The ASKIRCompilerService can call this.
 * If the project is a UI project it will be transformed to the version that can be compiled
 * in the frontend/handled in the backend.
 * @param locale locale to load the project
 * @param rootDir root of the project
 * @returnsa project corresponding to ASKIR for validation. The ASKIRCompilerService can call this
 */
async function loadASKIRProject(locale, rootDir = process.cwd()) {
    let projectConfig = undefined;
    const pkgJson = await (0, package_1.loadPackageConfig)(rootDir);
    if (pkgJson !== undefined) {
        projectConfig = getProjectConfig(rootDir, pkgJson);
    }
    if (!projectConfig) {
        const skillPackagePath = path_1.default.join(rootDir, "skill-package");
        const skillPackageConversationsPath = path_1.default.join(skillPackagePath, "conversations");
        const conversationsPath = path_1.default.join(rootDir, "conversations");
        let packageRootDir = undefined;
        if ((await (0, util_1.exists)(skillPackagePath)) && (await (0, util_1.exists)(skillPackageConversationsPath))) {
            packageRootDir = skillPackagePath;
        }
        else if (await (0, util_1.exists)(conversationsPath)) {
            packageRootDir = rootDir;
        }
        if (!packageRootDir) {
            throw new Error("Either package.json with ask section, ./skill-package/conversations or ./conversations folder must be present");
        }
        projectConfig = {
            srcDir: "conversations/" + locale + "/",
            name: "in-memory",
            rootDir: packageRootDir,
            outDir: "build",
            projectVersion: package_1.ProjectVersion.Beta1,
        };
    }
    let project = await (0, project_1.loadProject)(projectConfig, true);
    const uiSkillInformation = await (0, transformUIProject_1.getUISkillInformation)(project.sourceModules);
    if (!uiSkillInformation.isUISkill) {
        throw new Error("This is not a UI skill");
    }
    if (!new Set(uiSkillInformation.localesUsed).has(locale)) {
        throw new Error(`Locale ${locale} not found in skill locales: ${Array.from(uiSkillInformation.localesUsed.keys()).join(",")}`);
    }
    return (0, transformUIProject_1.transformUIProject)(project, uiSkillInformation.localesUsed, false);
}
exports.loadASKIRProject = loadASKIRProject;
function getProjectConfig(rootDir, packageConfig) {
    return {
        ...packageConfig.ask,
        name: packageConfig.name,
        rootDir,
    };
}
