// @alexa/acdl/dist/cjs/property-node.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.PropertyNode = void 0;
const _1 = require(".");
/**
 * Node in a tree useful to validate property paths in an utterance. See buildPropertyTree for details.
 */
class PropertyNode {
    /**
     *
     * @param property a Property in a type
     * @param propertyType the corresponding propertyType
     * @param children nodes corresponding to propertyType.Properties
     */
    constructor(property, propertyType, children = new Map()) {
        Object.defineProperty(this, "property", {
            enumerable: true,
            configurable: true,
            writable: true,
            value: property
        });
        Object.defineProperty(this, "propertyType", {
            enumerable: true,
            configurable: true,
            writable: true,
            value: propertyType
        });
        Object.defineProperty(this, "children", {
            enumerable: true,
            configurable: true,
            writable: true,
            value: children
        });
    }
    /**
     * Out of a string in Utterances<MealOrderType> like:
     *  'I want a {pizzaOrder.size} {pizzaOrder.pizzaType} pizza and a {drinkOrder.size} {drinkOrder.drinkType} drink'
     * This will build a tree like seen below where each Parenthesis is a Property with the given name and type:
     *
     *                                                    ('', MealOrderType)
     *
     *                 ('pizzaOrder', PizzaOrderType)                                      ('drinkOrder', DrinkOrderType)
     *
     *     ('pizzaType' ,PizzaType)          ('size', PizzaSizeType)             ('drinkType', DrinkType)        ('size', DrinkSizeType)
     *
     * In the process of building the tree 'no such property' errors can be produced.
     * With the tree, at each node, we can verify if the required properties have been specified
     * @param propertyPaths each ask.Name contains dot separate property path like 'pizzaOrder.size' plus its location for error messages.
     * @param type the type where all the first segments in propertyPaths are supposed to belong to.
     * @param context used for type's toString
     * @param getPropertyMap memoized type's getPropertyMap
     * @returns errors and the root of the tree
     */
    static buildPropertyTree(propertyPaths, type, context, getPropertyMap) {
        var _a, _b;
        const errors = [];
        const root = new PropertyNode({ name: "", type, kind: "Property", index: 0 }, type);
        for (const propertyPath of propertyPaths) {
            const segments = ((_a = propertyPath.name) !== null && _a !== void 0 ? _a : "").split(".");
            let currentType = type;
            let currentNode = root;
            let propertyFound = true;
            for (const segment of segments) {
                const propertyToTypes = getPropertyMap(currentType);
                const segmentProperty = propertyToTypes.get(segment);
                const segmentType = segmentProperty === null || segmentProperty === void 0 ? void 0 : segmentProperty.type;
                if (!segmentProperty) {
                    propertyFound = false;
                    // ignore if slot value is empty between braces as this will be caught later on during validation
                    if (segment.replace(/\s*/g, "").length > 0) {
                        errors.push(_1.ValidationErrorFactory.UnknownProperty({
                            expr: propertyPath,
                            attributes: {
                                propertyName: segment,
                                typeName: currentType.toString(context),
                            },
                        }));
                    }
                    continue;
                }
                if (!segmentType) {
                    propertyFound = false;
                    errors.push(_1.ValidationErrorFactory.NameNotFoundInUtteranceValue({
                        expr: propertyPath,
                        attributes: {
                            value: segment,
                        },
                    }));
                    continue;
                }
                let segmentNode = currentNode.children.get(segment);
                if (!segmentNode) {
                    segmentNode = new PropertyNode(segmentProperty, segmentType);
                    currentNode.children.set(segment, segmentNode);
                }
                currentType = segmentType.getNonOptionalType();
                currentNode = segmentNode;
            }
            const elementType = currentType.isList()
                ? currentType.getListItemType()
                : currentType.isOptional()
                    ? currentType.getNonOptionalType()
                    : currentType;
            const isTypeValid = (elementType === null || elementType === void 0 ? void 0 : elementType.isBuiltIn()) || (elementType === null || elementType === void 0 ? void 0 : elementType.isCoreType()) || (elementType === null || elementType === void 0 ? void 0 : elementType.isAlexaSchema()) || (elementType === null || elementType === void 0 ? void 0 : elementType.isSlotType());
            if (propertyFound && !(elementType === null || elementType === void 0 ? void 0 : elementType.isValidForUtterance())) {
                errors.push(_1.ValidationErrorFactory.UndefinedCatalogCustomType({
                    expr: propertyPath,
                    attributes: {
                        propertyName: (_b = propertyPath.name) !== null && _b !== void 0 ? _b : "",
                        propertyType: elementType === null || elementType === void 0 ? void 0 : elementType.toString(propertyPath),
                    },
                }));
            }
        }
        return { errors, root };
    }
}
exports.PropertyNode = PropertyNode;
