// @alexa/acdl/dist/cjs/resolve.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.resolveExpr = exports.resolveModule = exports.resolveProject = void 0;
const project_1 = require("./project");
const ask = __importStar(require("./ast"));
const visit_1 = require("./visit");
/**
 * Walks over all modules in the Project and performs all type and argument inference on the tree once to avoid repeating this cost later.
 *
 * @param project
 * @returns a project where all modules
 */
function resolveProject(project) {
    const checker = project.getTypeChecker();
    return new project_1.Project({
        config: project.config,
        external: project.external,
        source: {
            ...project.source,
            modules: project.sourceModules.map((mod) => ({
                ...mod,
                content: mod.content ? resolveModule(mod.content, checker) : undefined,
            })),
        },
    });
}
exports.resolveProject = resolveProject;
function resolveModule(mod, checker) {
    var _a;
    return new ask.Module(mod.uri, mod.schema, mod.namespace, mod.imports, mod.exports, (_a = mod.expressions) === null || _a === void 0 ? void 0 : _a.map((expr) => resolveExpr(expr, checker)));
}
exports.resolveModule = resolveModule;
function resolveExpr(expr, checker) {
    var _a, _b, _c, _d;
    if (expr.kind === "Call") {
        const apply = checker.getApply(expr);
        if (apply !== undefined) {
            const args = apply === null || apply === void 0 ? void 0 : apply.getArguments();
            // resolve the generic arguments
            const typeArgs = expr.genericArguments !== undefined
                ? (_a = expr.genericArguments) === null || _a === void 0 ? void 0 : _a.map((arg) => arg === null || arg === void 0 ? void 0 : arg.clone())
                : (_b = apply === null || apply === void 0 ? void 0 : apply.getGenericArguments()) === null || _b === void 0 ? void 0 : _b.map((type) => resolveTypeReference(type === null || type === void 0 ? void 0 : type.toTypeReference(), checker));
            const annotations = (_c = expr.annotations) === null || _c === void 0 ? void 0 : _c.map((annotation) => new ask.Annotation(resolveExpr(annotation.call, checker), annotation.loc));
            return new ask.Call((_d = expr.name) === null || _d === void 0 ? void 0 : _d.clone(), typeof expr.arguments === "string" ||
                typeof expr.arguments === "number" ||
                typeof expr.arguments === "boolean" ||
                expr.arguments === null ||
                expr.arguments === undefined
                ? expr.arguments
                : ask.isListLiteral(expr.arguments)
                    ? (0, visit_1.transformEachChild)(expr.arguments, (child) => resolveExpr(child, checker))
                    : typeof args === "object"
                        ? // TODO: is this preserving order and name locations
                            Object.entries(args).map(([name, arg], i) => new ask.Argument(new ask.Name(name), i, resolveExpr(arg.value, checker), arg.loc))
                        : undefined, typeArgs, annotations, expr.loc);
        }
    }
    const child = (0, visit_1.transformEachChild)(expr, (child) => resolveExpr(child, checker));
    return child;
}
exports.resolveExpr = resolveExpr;
function resolveTypeReference(type, checker) {
    var _a, _b, _c;
    if (type === undefined) {
        return undefined;
    }
    return new ask.TypeReference((_a = type.name) === null || _a === void 0 ? void 0 : _a.clone(), (_b = type.arguments) === null || _b === void 0 ? void 0 : _b.map((arg) => resolveTypeReference(arg, checker)), (_c = type.annotations) === null || _c === void 0 ? void 0 : _c.map((ann) => ann.clone()), undefined, undefined, // place the location of the resolved type arguments on the call's name
    true);
}
