// @alexa/acdl/dist/cjs/skill-context.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.loadSkillContext = void 0;
function loadSkillContext(skill) {
    var _a, _b, _c, _d, _e, _f, _g, _h, _j;
    const taskContextProjectionInfo = skill.getArgumentValueThing("taskContextProjectionInfo");
    if (taskContextProjectionInfo === null || taskContextProjectionInfo === void 0 ? void 0 : taskContextProjectionInfo.type.isTaskContextProjectionInfo()) {
        return {
            skill,
            loadContext: (_b = (_a = taskContextProjectionInfo.getProperty("loadContext")) === null || _a === void 0 ? void 0 : _a.asAction()) === null || _b === void 0 ? void 0 : _b.origin,
            loadContextThing: taskContextProjectionInfo.getProperty("loadContext"),
            loadContextDeclaration: (_c = taskContextProjectionInfo.apply) === null || _c === void 0 ? void 0 : _c.getArgumentValue("loadContext"),
            saveContext: (_e = (_d = taskContextProjectionInfo.getProperty("saveContext")) === null || _d === void 0 ? void 0 : _d.asAction()) === null || _e === void 0 ? void 0 : _e.origin,
            saveContextThing: taskContextProjectionInfo.getProperty("saveContext"),
            saveContextDeclaration: (_f = taskContextProjectionInfo.apply) === null || _f === void 0 ? void 0 : _f.getArgumentValue("saveContext"),
            trackActions: (_j = (_h = (_g = taskContextProjectionInfo
                .getProperty("trackActions")) === null || _g === void 0 ? void 0 : _g.asLiteralList()) === null || _h === void 0 ? void 0 : _h.getItems()) === null || _j === void 0 ? void 0 : _j.map((item) => { var _a; return (_a = item === null || item === void 0 ? void 0 : item.asAction()) === null || _a === void 0 ? void 0 : _a.origin; }).filter((a) => a !== undefined),
            trackActionsThing: taskContextProjectionInfo.getProperty("trackActions"),
            trackActionsDeclaration: skill.getArgumentValue("trackActions"),
        };
    }
    return undefined;
}
exports.loadSkillContext = loadSkillContext;
