// @alexa/acdl/dist/cjs/try.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.isRight = exports.isLeft = exports.right = exports.left = exports.TryMonad = exports.Try = exports.isTry = exports.isFailure = void 0;
function isFailure(a) {
    return a.kind === "Failure";
}
exports.isFailure = isFailure;
/**
 * Check if a value is an instance of `Try`.
 */
function isTry(a) {
    return a && typeof a.getOrThrow === "function";
}
exports.isTry = isTry;
/**
 * Monad for safely calling a function that can throw errors.
 *
 * @param get function so safely call
 * @param errorHandler optional handler for transforming errors
 */
function Try(get, errorHandler) {
    return new TryMonad(get, errorHandler);
}
exports.Try = Try;
/**
 * Encapsulates a computation that may throw errors.
 */
class TryMonad {
    constructor(thunk, errorHandler) {
        Object.defineProperty(this, "thunk", {
            enumerable: true,
            configurable: true,
            writable: true,
            value: thunk
        });
        Object.defineProperty(this, "errorHandler", {
            enumerable: true,
            configurable: true,
            writable: true,
            value: errorHandler
        });
    }
    map(f, err) {
        return Try(() => f(this.thunk()), err || this.errorHandler);
    }
    flatMap(f, err) {
        return Try(() => f(this.thunk()).getOrThrow(), err || this.errorHandler);
    }
    resolve(context) {
        try {
            return this.thunk();
        }
        catch (err) {
            if (this.errorHandler === undefined && err.code === undefined) {
                return undefined;
            }
            context.errors.push(...(this.errorHandler ? this.errorHandler(err) : [err]));
        }
        return undefined;
    }
    getOrDefault(t) {
        try {
            return this.getOrThrow();
        }
        catch (err) {
            return t;
        }
    }
    getOrThrow(error) {
        try {
            return this.thunk();
        }
        catch (err) {
            throw error || err;
        }
    }
}
exports.TryMonad = TryMonad;
const left = (left) => ({ left });
exports.left = left;
const right = (right) => ({ right });
exports.right = right;
const isLeft = (a) => "left" in a;
exports.isLeft = isLeft;
const isRight = (a) => "Right" in a;
exports.isRight = isRight;
