// @alexa/acdl/dist/cjs/metrics/metric-client.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.MetricClient = exports.MetricAction = exports.MetricActionResult = void 0;
const uuid_1 = require("uuid");
const axios_1 = __importDefault(require("axios"));
const ProfileNameGetter_1 = require("../acc/ProfileNameGetter");
const METRICS = {
    ENDPOINT: "https://client-telemetry.amazonalexa.com",
    STOREABLE_KEYS: ["skill_id", "locale", "template_url", "stage"],
};
exports.MetricActionResult = {
    SUCCESS: "Success",
    FAILURE: "Failure",
};
/**
 * Metric action includes the name and results of CLI command.
 */
class MetricAction {
    /**
     * @constructor
     * @param name - The action name.
     * @param type - The action type.
     */
    constructor(name, type) {
        Object.defineProperty(this, "name", {
            enumerable: true,
            configurable: true,
            writable: true,
            value: void 0
        });
        Object.defineProperty(this, "type", {
            enumerable: true,
            configurable: true,
            writable: true,
            value: void 0
        });
        Object.defineProperty(this, "startTime", {
            enumerable: true,
            configurable: true,
            writable: true,
            value: void 0
        });
        Object.defineProperty(this, "endTime", {
            enumerable: true,
            configurable: true,
            writable: true,
            value: void 0
        });
        Object.defineProperty(this, "result", {
            enumerable: true,
            configurable: true,
            writable: true,
            value: void 0
        });
        Object.defineProperty(this, "failureMessage", {
            enumerable: true,
            configurable: true,
            writable: true,
            value: void 0
        });
        Object.defineProperty(this, "id", {
            enumerable: true,
            configurable: true,
            writable: true,
            value: void 0
        });
        Object.defineProperty(this, "_ended", {
            enumerable: true,
            configurable: true,
            writable: true,
            value: void 0
        });
        Object.defineProperty(this, "options", {
            enumerable: true,
            configurable: true,
            writable: true,
            value: void 0
        });
        Object.defineProperty(this, "optionData", {
            enumerable: true,
            configurable: true,
            writable: true,
            value: void 0
        });
        this.endTime = null;
        this.failureMessage = "";
        this.name = name;
        this.result = null;
        this.startTime = new Date();
        this.type = type;
        this.id = (0, uuid_1.v4)();
        this.options = [];
        this.optionData = {};
        this._ended = false;
    }
    /**
     * Add option to list, store value only if in {@link METRICS.STOREABLE_KEYS}
     * @param optionName name of the option
     * @param optionValue value of the option
     */
    setOption(optionName, optionValue) {
        const schemaOption = optionName.split("-").join("_");
        this.options.push(schemaOption);
        if (METRICS.STOREABLE_KEYS.includes(schemaOption)) {
            this.optionData[schemaOption] = optionValue;
        }
    }
    /**
     * Closes action
     * @param error Error object or string indicating error.
     */
    end(error = null) {
        if (this._ended)
            return;
        // if Error object extract error message,
        // otherwise error message string or null was passed as a parameter
        const errorMessage = error && error instanceof Error ? error.message : error;
        this.result = errorMessage ? exports.MetricActionResult.FAILURE : exports.MetricActionResult.SUCCESS;
        this.failureMessage = errorMessage || "";
        this.endTime = new Date();
        this._ended = true;
    }
    /**
     * Implementation of custom toJSON method to modify serialization with JSON.stringify
     */
    toJSON() {
        return {
            end_time: this.endTime,
            failure_message: this.failureMessage,
            name: this.name,
            result: this.result,
            start_time: this.startTime,
            type: this.type,
            id: this.id,
            options: this.options,
            ...this.optionData,
        };
    }
}
exports.MetricAction = MetricAction;
/**
 * Client that communicates with telemetry endpoint.
 */
class MetricClient {
    /**
     * @constructor
     */
    constructor(currentNpmVersion, askCliConfig) {
        Object.defineProperty(this, "httpClient", {
            enumerable: true,
            configurable: true,
            writable: true,
            value: void 0
        });
        Object.defineProperty(this, "serverUrl", {
            enumerable: true,
            configurable: true,
            writable: true,
            value: void 0
        });
        Object.defineProperty(this, "postRetries", {
            enumerable: true,
            configurable: true,
            writable: true,
            value: void 0
        });
        Object.defineProperty(this, "enabled", {
            enumerable: true,
            configurable: true,
            writable: true,
            value: void 0
        });
        Object.defineProperty(this, "data", {
            enumerable: true,
            configurable: true,
            writable: true,
            value: void 0
        });
        Object.defineProperty(this, "askCliConfig", {
            enumerable: true,
            configurable: true,
            writable: true,
            value: void 0
        });
        this.httpClient = axios_1.default.create({
            timeout: 3000,
            headers: { "Content-Type": "text/plain" },
        });
        this.serverUrl = METRICS.ENDPOINT;
        this.postRetries = 3;
        this.askCliConfig = askCliConfig;
        this.enabled = this.isEnabled();
        // initialize data
        this.data = {
            version: currentNpmVersion,
            machineId: this.getMachineId(),
            timeStarted: new Date(),
            newUser: false,
            timeUploaded: null,
            clientId: "acc",
            actions: [],
        };
    }
    /**
     * Starts action
     * @param name - The action name
     * @param type - The action type
     * @return the metric action that was started, {@link MetricAction}
     */
    startAction(name, type) {
        const action = new MetricAction(name, type);
        this.data.actions.push(action);
        return action;
    }
    /**
     * Set option for the most recently started action
     * @param optionName name of the option
     * @param optionValue value of the option
     */
    setOption(optionName, optionValue) {
        const actions = this.data.actions;
        if (actions.length) {
            actions[actions.length - 1].setOption(optionName, optionValue);
        }
    }
    /**
     * Returns current data store in the metric client
     * @return the metric data, {@link MetricClientData}
     */
    getData() {
        return this.data;
    }
    /**
     * Sends data to the metric server
     * @param error - Error object or string indicating error.
     * @returns whether data was sent successfully
     */
    async sendData(error = null) {
        if (!this.enabled) {
            this.data.actions = [];
            return { success: true };
        }
        this.data.actions.forEach((action) => action.end(error));
        try {
            await this.upload();
            this.data.actions = [];
            return { success: true };
        }
        catch (_a) {
            return { success: false };
        }
    }
    /**
     * Implementation of custom toJSON method to modify serialization with JSON.stringify
     */
    toJSON() {
        return {
            version: this.data.version,
            machine_id: this.data.machineId,
            time_started: this.data.timeStarted,
            new_user: this.data.newUser,
            time_uploaded: this.data.timeUploaded,
            client_id: this.data.clientId,
            actions: this.data.actions,
        };
    }
    async upload() {
        this.data.timeUploaded = new Date();
        const payload = JSON.stringify({ payload: this });
        const postPromise = () => this.httpClient.post(this.serverUrl, payload);
        await this.retry(this.postRetries, postPromise);
    }
    async retry(retries, func) {
        try {
            await func();
        }
        catch (error) {
            if (retries == 1)
                throw error;
            await this.retry(retries - 1, func);
        }
    }
    isEnabled() {
        if (ProfileNameGetter_1.ProfileNameGetter.isEnvProfile())
            return true;
        if (process.env.ASK_SHARE_USAGE === "false")
            return false;
        if (!this.askCliConfig.hasConfig())
            return false;
        return this.askCliConfig.getShareUsage() && !!this.getMachineId();
    }
    /**
     * gets the machine id from ~/.ask/cli_config.
     * Special cases:
     *    1) If ~/.ask/cli_config was not found or could not be parsed this.enabled will be false and
     * no machine id is retrieved since no metrics will be sent
     *    2) If ~/.ask/cli_config was found, but has no machineId we will:
     *         a) write a new uuid to ~/.ask/cli_config if possible
     *         b) diaable metrics if we cannot write to ~/.ask/cli_config
     * @returns a machine id from ~/.ask/cli_config
     */
    getMachineId() {
        if (!this.askCliConfig.hasConfig())
            return "";
        if (ProfileNameGetter_1.ProfileNameGetter.isEnvProfile())
            return "all_environmental";
        let machineId = this.askCliConfig.getMachineId();
        if (machineId) {
            return machineId;
        }
        return this.askCliConfig.createMachineId();
    }
}
exports.MetricClient = MetricClient;
