// @alexa/acdl/dist/esm/ask-cli-config.js
import * as os from "os";
import * as fs from "fs-extra";
import { v4 as uuid } from "uuid";
export class AskCliConfig {
    /**
     * Load the user's ASK CLI Config.
     *
     * @param path path of the cli_config - defaults to `~/.ask/cli_config`
     */
    constructor() {
        Object.defineProperty(this, "askCliConfig", {
            enumerable: true,
            configurable: true,
            writable: true,
            value: void 0
        });
        Object.defineProperty(this, "cliConfigPath", {
            enumerable: true,
            configurable: true,
            writable: true,
            value: void 0
        });
        Object.defineProperty(this, "errorLoadingAskCliConfig", {
            enumerable: true,
            configurable: true,
            writable: true,
            value: void 0
        });
        this.cliConfigPath = os.homedir() + "/.ask/cli_config";
        let cliConfigString;
        try {
            cliConfigString = fs.readFileSync(this.cliConfigPath).toString("utf8");
        }
        catch (err) {
            this.errorLoadingAskCliConfig = `cannot read config file: ${this.cliConfigPath}. Run 'ask configure'.`;
            return;
        }
        try {
            this.askCliConfig = JSON.parse(cliConfigString);
        }
        catch (err) {
            this.errorLoadingAskCliConfig = `failed to parse config file: ${this.cliConfigPath}`;
            return;
        }
    }
    hasConfig() {
        return !!this.askCliConfig;
    }
    createMachineId() {
        const machine_id = uuid();
        try {
            fs.writeFileSync(this.cliConfigPath, JSON.stringify({ ...(this.askCliConfig || {}), machine_id }, null, 2), "utf-8");
        }
        catch (error) {
            return "";
        }
        return machine_id;
    }
    getLwaProfile(profileName) {
        if (this.errorLoadingAskCliConfig) {
            throw new Error(this.errorLoadingAskCliConfig);
        }
        const lwaProfile = this.askCliConfig.profiles[profileName];
        if (lwaProfile === undefined) {
            this.errorLoadingAskCliConfig = `could not load profile '${profileName}' from ${this.cliConfigPath}`;
        }
        return lwaProfile;
    }
    getShareUsage() {
        const shareUsage = this.askCliConfig.share_usage;
        if (shareUsage !== undefined)
            return !!shareUsage;
        return true;
    }
    getMachineId() {
        return this.askCliConfig.machine_id;
    }
}
