// @alexa/acdl/dist/esm/constants.js
import { AlexaConversations } from "./ast";
export var Constants;
(function (Constants) {
    Constants.DEFAULT_CONFIG_FILE_NAME = "ask.json";
    Constants.RESERVED_PREFIXES = [
        "AMAZON",
        // TODO: add schema and conversations back to this list once we figure out how to give exception to @alexa/schema
        // issue link: https://github.com/alexa/ask-ac/issues/92
        // 'com.amazon.alexa.schema',
        // 'com.amazon.alexa.ask.conversations',
        "slotTypes",
        "prompts",
    ];
    Constants.BINARY_OPERATOR_ACTIONS_SET = new Set([
        AlexaConversations.And,
        AlexaConversations.Eq,
        AlexaConversations.Gt,
        AlexaConversations.Gte,
        AlexaConversations.in,
        AlexaConversations.Lt,
        AlexaConversations.Lte,
        AlexaConversations.Neq,
        AlexaConversations.Or,
    ]);
    Constants.OPERATOR_ACTIONS_SET = new Set([
        // Logical operations
        AlexaConversations.And,
        AlexaConversations.Eq,
        AlexaConversations.Gt,
        AlexaConversations.Gte,
        AlexaConversations.in,
        AlexaConversations.Lt,
        AlexaConversations.Lte,
        AlexaConversations.Neq,
        AlexaConversations.Or,
        AlexaConversations.in,
        AlexaConversations.is,
        AlexaConversations.isInCatalog,
        AlexaConversations.Not,
    ]);
    Constants.FILE_ACCESS_ACTIONS_SET = new Set([
        // Logical operations
        AlexaConversations.apl,
        AlexaConversations.apla,
    ]);
    /**
     * Name of the Skill Manifest at the root of a Skill Package.
     */
    Constants.SkillManifestFileName = "skill.json";
    /**
     * Name of the Skill Package Folder.
     */
    Constants.SkillPackageFolderName = "skill-package";
    /**
     * Name of the folder at the root of a Skill Package containing the Interaction Models.
     */
    Constants.InteractionsModelFolderName = "interactionModels";
    /**
     * Name of the folder under ${InteractionsModelFolderName} containing the custom Interaction Model JSON Files.
     */
    Constants.CustomFolderName = "custom";
    /**
     * Name of the folder inside the ${InteractionModelFolderName} containing the Alexa Conversations ASKIR files.
     */
    Constants.ConversationsFolderName = "conversations";
    /**
     * Name of the folder containing Responses.
     */
    Constants.ResponseFolderName = "response";
    /**
     * Name of the folder containing APL display responses.
     */
    Constants.DisplayFolderName = "display";
    /**
     * Name of the folder containing APLA prompt responses.
     */
    Constants.PromptsFolderName = "prompts";
    /**
     * Name of the folder containing Prompts.
     */
    Constants.AssetsFolderName = "assets";
    /**
     * Name of the folder containing Data Store Packages.
     */
    Constants.DataStorePackagesFolderName = "_dataStorePackages";
    /**
     * Name of the folder containing Prompts.
     */
    Constants.TasksFolderName = "tasks";
    /**
     * Name of the folder containing the In Skill Product files.
     */
    Constants.InSkillPurchaseFolderName = "isps";
    /**
     * File Extension for ACDL Files.
     */
    Constants.ACDLExtension = "acdl";
    /**
     * File Extension for ASKIR JSON Files.
     */
    Constants.ASKIRJsonExtension = "json";
    /**
     * Name of the NPM Package containing Alexa Schema.
     */
    Constants.AlexaSchemaPackageName = "@alexa/schema";
    /**
     * Name of the NPM Package containing Alexa Conversations Core Library.
     */
    Constants.ACCorePackageName = "@alexa/ac-core";
    /**
     * Name of the folder containing ASKIR Dependencies.
     */
    Constants.DependenciesFolderName = "_dependencies";
    /**
     * Name of the folder containing Compiled ASKIR Files in a `_dependencies` folder.
     */
    Constants.CompiledOutputFoldername = "compiled-output";
    /**
     * Name of the file for APLA and APL
     */
    Constants.APLADocumentFileName = "document.json";
    /**
     * The maximum size of IR is 4mb
     */
    Constants.MaximumFileSize = 4 * 1024 * 1024;
    Constants.BuiltinIntentPrefix = "AMAZON.";
})(Constants || (Constants = {}));
/**
 * Enum of the allowed `type` field in an APL/A file.
 *
 * Both `APL-A` and `APLA` are allowed for APL-A files.
 *
 * Only `APL` is allowed for APL files.
 */
export var APLFileType;
(function (APLFileType) {
    APLFileType["APLA"] = "APLA";
    APLFileType["APL_A"] = "APL-A";
    APLFileType["APL"] = "APL";
})(APLFileType || (APLFileType = {}));
export const APLAFileTypes = [APLFileType.APLA, APLFileType.APL_A];
export const APLFileTypes = [APLFileType.APL];
/**
 * Contains reusable error messages used for validations.
 */
export var ValidationErrorMessages;
(function (ValidationErrorMessages) {
    ValidationErrorMessages["MustBeBackedByCatalog"] = "must be backed by a catalog - choose a type from ask.interactionModel.slotTypes.*, com.amazon.ask.types.builtins.AMAZON.*, or annotate your type with the @catalog annotation.";
})(ValidationErrorMessages || (ValidationErrorMessages = {}));
export var GeneratedInteractionModelModule;
(function (GeneratedInteractionModelModule) {
    /**
     * The root acdl namespace for each each generated interaction module.
     */
    GeneratedInteractionModelModule.namespacePrefix = "ask.interactionModel";
    /**
     * The uri prefix that each generated module URI will start with.
     */
    GeneratedInteractionModelModule.uriPrefix = "generated-im";
})(GeneratedInteractionModelModule || (GeneratedInteractionModelModule = {}));
export var Namespaces;
(function (Namespaces) {
    /**
     * Namespace for generated slotTypes
     */
    Namespaces.slotTypes = "ask.interactionModel.slotTypes";
    /**
     * Namespace for generated catalogs
     */
    Namespaces.catalogs = "ask.interactionModel.catalogs";
    /**
     * Namespace for generated intents
     */
    Namespaces.intents = "ask.interactionModel.intents";
    /**
     * Namespace for generated intent types
     */
    Namespaces.intentTypes = "ask.interactionModel.intentTypes";
    /**
     * Namespace given to APL/A prompts loaded from `response/prompts`.
     */
    Namespaces.prompts = "prompts";
    /**
     * Namespace given to APL display templates loaded from `response/display`.
     */
    Namespaces.display = "displays";
    /**
     * Namespace given to builtin types
     */
    Namespaces.builtinTypes = "com.amazon.ask.types.builtins.AMAZON";
    /**
     * Namespace given to builtin catalogs
     */
    Namespaces.builtinCatalogs = "com.amazon.ask.catalogs.builtins.AMAZON";
})(Namespaces || (Namespaces = {}));
