// @alexa/acdl/dist/esm/error.js
import path from "path";
import { ValidationErrorFactory, getExpressionIfLocField } from "./error-factory";
/**
 * Used for errors in non ACDL files, most likely JSON files
 */
export class FileError extends Error {
    constructor(message, uri) {
        super(message);
        Object.defineProperty(this, "message", {
            enumerable: true,
            configurable: true,
            writable: true,
            value: message
        });
        Object.defineProperty(this, "uri", {
            enumerable: true,
            configurable: true,
            writable: true,
            value: uri
        });
        this.uri = uri;
        this.message = message;
    }
}
export function logProjectErrors(errors) {
    errors.forEach(logProjectError);
}
export function logProjectWarnings(errors) {
    if (errors !== undefined) {
        errors.filter((error) => error.code.category === "Warning" /* ErrorCategory.Warning */).forEach(logProjectError);
    }
}
export function logProjectError(error) {
    console.error(stringifyProjectError(error));
}
/**
 * Gets a string out of an error if it is a ProjectError or a regular Javascript Error
 */
export function stringifyError(error) {
    var _a;
    const errorAny = error;
    if ((_a = errorAny.code) === null || _a === void 0 ? void 0 : _a.category) {
        return stringifyProjectError(error);
    }
    return error.message;
}
export function stringifyProjectError(error) {
    const errorStringArray = [];
    if (error.code.category === "Warning" /* ErrorCategory.Warning */) {
        errorStringArray.push(error.code.category);
    }
    if (error.loc && error.uri) {
        errorStringArray.push(stringifyLocation(error.uri, error.loc));
    }
    errorStringArray.push(error.message);
    return errorStringArray.join(": ");
}
export function stringifyLocation(errorUri, loc) {
    // print the error relative to the current working directory so that links in the terminal link directly to the file
    const uri = path.isAbsolute(errorUri) ? path.relative(process.cwd(), errorUri) : errorUri;
    if (loc) {
        return `${uri}(${loc.begin.line},${loc.begin.character})`;
    }
    return uri;
}
export function filterDuplicateErrors(errors) {
    const seen = new Set();
    const e = errors.filter((err) => {
        if (err === undefined) {
            return false;
        }
        const str = `${err.message}${JSON.stringify(err.loc)}`;
        if (seen.has(str)) {
            return false;
        }
        seen.add(str);
        return true;
    });
    return e;
}
/**
 * Sort error array based on uri and loc
 * error without uri directly compare the loc
 * error with uri first compare uri then compare the loc
 * when loc are identical, compare the error message
 * @param errors
 * @returns
 */
export function sortErrors(errors) {
    return errors.sort((a, b) => {
        if (a.uri === undefined && b.uri === undefined) {
            const result = compareLoc(a.loc, b.loc);
            return result === 0 ? compareMessage(a.message, b.message) : result;
        }
        if (a.uri === undefined) {
            return -1;
        }
        if (b.uri === undefined) {
            return 1;
        }
        if (a.uri > b.uri) {
            return 1;
        }
        if (a.uri < b.uri) {
            return -1;
        }
        const result = compareLoc(a.loc, b.loc);
        return result === 0 ? compareMessage(a.message, b.message) : result;
    });
    /**
     * compare two location
     * error without loc is placed before other errors
     * when loc is exist, compare the start and then end
     * @param firstLoc
     * @param secondLoc
     * @returns
     */
    function compareLoc(firstLoc, secondLoc) {
        if (firstLoc === undefined && secondLoc === undefined) {
            return -1;
        }
        if (firstLoc === undefined) {
            return -1;
        }
        if (secondLoc === undefined) {
            return 1;
        }
        return firstLoc.begin.line !== secondLoc.begin.line
            ? firstLoc.begin.line - secondLoc.begin.line
            : firstLoc.begin.character !== secondLoc.begin.character
                ? firstLoc.begin.character - secondLoc.begin.character
                : firstLoc.end.line !== secondLoc.end.line
                    ? firstLoc.end.line - secondLoc.end.line
                    : firstLoc.end.character - secondLoc.end.character;
    }
    function compareMessage(firstMessage, secondMessage) {
        if (firstMessage <= secondMessage) {
            return -1;
        }
        return 1;
    }
}
export function typeError(actual, expected, scope, errorExpr) {
    return ValidationErrorFactory.TypeError({
        expr: getExpressionIfLocField(errorExpr, scope),
        attributes: {
            actualType: actual.toString(scope),
            expectedType: typeof expected === "string" ? expected : expected.toString(scope),
        },
    });
}
export function nameNotFound(name, loc, node) {
    var _a, _b;
    const fqn = name.name ? [name.name] : [];
    let context = node === null || node === void 0 ? void 0 : node.context;
    while ((context === null || context === void 0 ? void 0 : context.kind) === "PropertyReference") {
        if (((_a = context.name) === null || _a === void 0 ? void 0 : _a.name) === undefined) {
            break;
        }
        fqn.push((_b = context.name) === null || _b === void 0 ? void 0 : _b.name);
        context = context.context;
    }
    return ValidationErrorFactory.NameNotFound({
        expr: name,
        attributes: {
            name: fqn.join("."),
        },
    });
}
export function hasTerminalErrors(errors) {
    return errors.some((error) => isTerminalError(error));
}
function isTerminalError(error) {
    return error.code.category !== "Warning" /* ErrorCategory.Warning */;
}
