// @alexa/acdl/dist/esm/project-config.js
import path from "path";
import { loadPackageConfig, ProjectVersion, loadPackageConfigSync } from "./package";
import { loadProject } from "./project";
import { getUISkillInformation, transformUIProject } from "./transformUIProject";
import { exists, existsSync, readJsonFile, readJsonFileSync } from "./util";
/**
 * Load a project from a directory on disk.
 *
 * We first look for a `package.json` with an `ask` section or an `ask.json` file at the `rootDir` and use
 * it as configuration if it exists.
 *
 * Otherwise we then check if there an `ask-resources.json` file pointing to a `skill-package` folder. Failing that,
 * we then look for a `skill-package` folder at the `rootDir`. In all of these cases, we default to a `1.0.0-beta1`
 * project configuration.
 *
 * @param rootDir path of the project containing `package.json` or `ask.json` configuration file.
 * @param profile ASK CLI profile to use when inspecting `ask-resources.json`.
 */
export function loadProjectConfigSync(rootDir = process.cwd(), profile = "default") {
    var _a, _b, _c, _d;
    const pkgJson = loadPackageConfigSync(rootDir);
    if (pkgJson !== undefined) {
        return getProjectConfig(rootDir, pkgJson);
    }
    const askResources = readJsonFileSync(path.join(rootDir, "ask-resources.json"));
    const skillPackagePath = path.join(rootDir, (_d = (_c = (_b = (_a = askResources === null || askResources === void 0 ? void 0 : askResources.profiles) === null || _a === void 0 ? void 0 : _a[profile]) === null || _b === void 0 ? void 0 : _b.skillMetadata) === null || _c === void 0 ? void 0 : _c.src) !== null && _d !== void 0 ? _d : "skill-package");
    if (existsSync(skillPackagePath)) {
        return {
            rootDir,
            srcDir: "conversations",
            outDir: "build",
            projectVersion: ProjectVersion.Beta1,
            sourceRootDir: skillPackagePath,
        };
    }
    throw new Error(`expected 'package.json', 'ask-resources.json' or 'skill-package/' to exist at location: '${rootDir}'`);
}
/**
 * Load a project from a directory on disk.
 *
 * We first look for a `package.json` with an `ask` section or an `ask.json` file at the `rootDir` and use
 * it as configuration if it exists.
 *
 * Otherwise we then check if there an `ask-resources.json` file pointing to a `skill-package` folder. Failing that,
 * we then look for a `skill-package` folder at the `rootDir`. In all of these cases, we default to a `1.0.0-beta1`
 * project configuration.
 *
 * @param rootDir path of the project containing `package.json` or `ask.json` configuration file.
 * @param profile ASK CLI profile to use when inspecting `ask-resources.json`.
 */
export async function loadProjectConfig(rootDir = process.cwd(), profile = "default") {
    var _a, _b, _c, _d;
    const pkgJson = await loadPackageConfig(rootDir);
    if (pkgJson !== undefined) {
        return getProjectConfig(rootDir, pkgJson);
    }
    const askResources = await readJsonFile(path.join(rootDir, "ask-resources.json"));
    const skillPackagePath = path.join(rootDir, (_d = (_c = (_b = (_a = askResources === null || askResources === void 0 ? void 0 : askResources.profiles) === null || _a === void 0 ? void 0 : _a[profile]) === null || _b === void 0 ? void 0 : _b.skillMetadata) === null || _c === void 0 ? void 0 : _c.src) !== null && _d !== void 0 ? _d : "skill-package");
    if (await exists(skillPackagePath)) {
        return {
            rootDir,
            srcDir: "conversations",
            outDir: "build",
            projectVersion: ProjectVersion.Beta1,
            sourceRootDir: skillPackagePath,
        };
    }
    throw new Error(`expected 'package.json', 'ask-resources.json' or 'skill-package/' to exist at location: '${rootDir}'`);
}
/**
 * Gets a project corresponding to ASKIR for validation. The ASKIRCompilerService can call this.
 * If the project is a UI project it will be transformed to the version that can be compiled
 * in the frontend/handled in the backend.
 * @param locale locale to load the project
 * @param rootDir root of the project
 * @returnsa project corresponding to ASKIR for validation. The ASKIRCompilerService can call this
 */
export async function loadASKIRProject(locale, rootDir = process.cwd()) {
    let projectConfig = undefined;
    const pkgJson = await loadPackageConfig(rootDir);
    if (pkgJson !== undefined) {
        projectConfig = getProjectConfig(rootDir, pkgJson);
    }
    if (!projectConfig) {
        const skillPackagePath = path.join(rootDir, "skill-package");
        const skillPackageConversationsPath = path.join(skillPackagePath, "conversations");
        const conversationsPath = path.join(rootDir, "conversations");
        let packageRootDir = undefined;
        if ((await exists(skillPackagePath)) && (await exists(skillPackageConversationsPath))) {
            packageRootDir = skillPackagePath;
        }
        else if (await exists(conversationsPath)) {
            packageRootDir = rootDir;
        }
        if (!packageRootDir) {
            throw new Error("Either package.json with ask section, ./skill-package/conversations or ./conversations folder must be present");
        }
        projectConfig = {
            srcDir: "conversations/" + locale + "/",
            name: "in-memory",
            rootDir: packageRootDir,
            outDir: "build",
            projectVersion: ProjectVersion.Beta1,
        };
    }
    let project = await loadProject(projectConfig, true);
    const uiSkillInformation = await getUISkillInformation(project.sourceModules);
    if (!uiSkillInformation.isUISkill) {
        throw new Error("This is not a UI skill");
    }
    if (!new Set(uiSkillInformation.localesUsed).has(locale)) {
        throw new Error(`Locale ${locale} not found in skill locales: ${Array.from(uiSkillInformation.localesUsed.keys()).join(",")}`);
    }
    return transformUIProject(project, uiSkillInformation.localesUsed, false);
}
function getProjectConfig(rootDir, packageConfig) {
    return {
        ...packageConfig.ask,
        name: packageConfig.name,
        rootDir,
    };
}
