// @alexa/acdl/dist/esm/try.js
export function isFailure(a) {
    return a.kind === "Failure";
}
/**
 * Check if a value is an instance of `Try`.
 */
export function isTry(a) {
    return a && typeof a.getOrThrow === "function";
}
/**
 * Monad for safely calling a function that can throw errors.
 *
 * @param get function so safely call
 * @param errorHandler optional handler for transforming errors
 */
export function Try(get, errorHandler) {
    return new TryMonad(get, errorHandler);
}
/**
 * Encapsulates a computation that may throw errors.
 */
export class TryMonad {
    constructor(thunk, errorHandler) {
        Object.defineProperty(this, "thunk", {
            enumerable: true,
            configurable: true,
            writable: true,
            value: thunk
        });
        Object.defineProperty(this, "errorHandler", {
            enumerable: true,
            configurable: true,
            writable: true,
            value: errorHandler
        });
    }
    map(f, err) {
        return Try(() => f(this.thunk()), err || this.errorHandler);
    }
    flatMap(f, err) {
        return Try(() => f(this.thunk()).getOrThrow(), err || this.errorHandler);
    }
    resolve(context) {
        try {
            return this.thunk();
        }
        catch (err) {
            if (this.errorHandler === undefined && err.code === undefined) {
                return undefined;
            }
            context.errors.push(...(this.errorHandler ? this.errorHandler(err) : [err]));
        }
        return undefined;
    }
    getOrDefault(t) {
        try {
            return this.getOrThrow();
        }
        catch (err) {
            return t;
        }
    }
    getOrThrow(error) {
        try {
            return this.thunk();
        }
        catch (err) {
            throw error || err;
        }
    }
}
export const left = (left) => ({ left });
export const right = (right) => ({ right });
export const isLeft = (a) => "left" in a;
export const isRight = (a) => "Right" in a;
